<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>bean count[-samples]: Count (reporter) screen data &#8212; bean 1.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=4f649999" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css?v=039e1c02" />
    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js?v=af2ce170"></script>
    <script src="../_static/doctools.js?v=888ff710"></script>
    <script src="../_static/sphinx_highlight.js?v=4825356b"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="bean-count-samples-count-reporter-screen-data">
<h1><code class="docutils literal notranslate"><span class="pre">bean</span> <span class="pre">count[-samples]</span></code>: Count (reporter) screen data<a class="headerlink" href="#bean-count-samples-count-reporter-screen-data" title="Permalink to this heading">¶</a></h1>
<p><code class="docutils literal notranslate"><span class="pre">bean</span> <span class="pre">count-samples</span></code> (or <code class="docutils literal notranslate"><span class="pre">bean</span> <span class="pre">count</span></code> for a single sample) maps guide into guide counts, <strong>allowing for base transition in spacer sequence</strong>. When the matched reporter information is provided, it can count the <strong>target site edits</strong> and <strong>alleles produced by each guide</strong>. Mapping is efficiently done based on <a class="reference external" href="https://github.com/pinellolab/CRISPResso2">CRISPResso2</a> modified for base-edit-aware mapping.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>bean count-samples \
  --input sample_list.csv   `# sample with lines &#39;R1_filepath,R2_filepath,sample_name\n&#39;` \
  -b A                      `# base that is being edited (A/G)` \
  -f sgRNA_info_table.csv   `# sgRNA information` \
  -o .                      `# output directory` \
  -r                        `# read edit/allele information from reporter` \
  -t 12                     `# number of threads` \
  --name my_sorting_screen  `# name of this sample run` \
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">bean</span> <span class="n">count</span> <span class="o">--</span><span class="n">R1</span> <span class="n">R1</span><span class="o">.</span><span class="n">fq</span> <span class="o">--</span><span class="n">R2</span> <span class="n">R2</span><span class="o">.</span><span class="n">fq</span> <span class="o">-</span><span class="n">b</span> <span class="n">A</span> <span class="o">-</span><span class="n">f</span> <span class="n">sgRNA_info_table</span><span class="o">.</span><span class="n">csv</span> <span class="o">-</span><span class="n">r</span>
</pre></div>
</div>
<p>By default, <code class="docutils literal notranslate"><span class="pre">bean</span> <span class="pre">count[-samples]</span></code> assume R1 and R2 are trimmed off of the adapter sequence. You may need to adjust the command arguments according to your read structure.</p>
<blockquote>
<div><p><span class="raw-html-m2r"><img src="../../imgs/sequence_struct.png" alt="Read structuren" width="600"/></span></p>
</div></blockquote>
<p>See full detail <a class="reference external" href="#full-parameters">below</a>.</p>
</section>
<section id="input-file-format">
<h1>Input file format<a class="headerlink" href="#input-file-format" title="Permalink to this heading">¶</a></h1>
<p>See <a class="reference internal" href="../input.html#input"><span class="std std-ref">Input file format</span></a> for input file formats.</p>
</section>
<section id="output-file-format">
<h1>Output file format<a class="headerlink" href="#output-file-format" title="Permalink to this heading">¶</a></h1>
<p><code class="docutils literal notranslate"><span class="pre">count</span></code> or <code class="docutils literal notranslate"><span class="pre">count-samples</span></code> produces <code class="docutils literal notranslate"><span class="pre">.h5ad</span></code> and <code class="docutils literal notranslate"><span class="pre">.xlsx</span></code> file with guide and per-guide allele counts.</p>
<ul class="simple">
<li><p><cite>.h5ad</cite>: This output file follows annotated matrix format compatible with <cite>AnnData</cite> and is based on <cite>Screen</cite> object in [purturb_tools](<a class="reference external" href="https://github.com/pinellolab/perturb-tools">https://github.com/pinellolab/perturb-tools</a>). See <a class="reference external" href="#data-structure">Data Structure</a> section for more information.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">.xlsx</span></code>: This output file contains <code class="docutils literal notranslate"><span class="pre">.guides</span></code>, <code class="docutils literal notranslate"><span class="pre">.samples</span></code>, <code class="docutils literal notranslate"><span class="pre">.X[_bcmatch,_edits]</span></code>. (<code class="docutils literal notranslate"><span class="pre">allele_tables</span></code> are often too large to write into an Excel!)</p></li>
</ul>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">bean</a></h1>








<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../gwas.html">GWAS variant library</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cds.html">Coding sequence tiling library</a></li>
<li class="toctree-l1"><a class="reference internal" href="../input.html">Input file format</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../subcommands.html">Subcommands</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../ReporterScreen_api.html">ReporterScreen API tutorial</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2024, Jayoung Ryu.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 7.1.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.13</a>
      
      |
      <a href="../_sources/commands/count.md.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>