<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>filter: Filtering (and optionally translating) alleles &#8212; bean 1.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=d1102ebc" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css?v=12dfc556" />
    <script src="../_static/documentation_options.js?v=8d563738"></script>
    <script src="../_static/doctools.js?v=888ff710"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="filter-filtering-and-optionally-translating-alleles">
<h1><code class="docutils literal notranslate"><span class="pre">filter</span></code>: Filtering (and optionally translating) alleles<a class="headerlink" href="#filter-filtering-and-optionally-translating-alleles" title="Link to this heading">¶</a></h1>
<p>As <code class="docutils literal notranslate"><span class="pre">tiling</span></code> mode of <code class="docutils literal notranslate"><span class="pre">bean</span> <span class="pre">run</span></code> accounts for any robustly observed alleles, <code class="docutils literal notranslate"><span class="pre">bean</span> <span class="pre">filter</span></code> filters for such alleles.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>bean<span class="w"> </span>filter<span class="w"> </span>my_sorting_screen_masked.h5ad<span class="w"> </span><span class="se">\</span>
-o<span class="w"> </span>my_sorting_screen_filtered.h5ad<span class="w">  </span><span class="sb">`</span><span class="c1"># Output file path` \</span>
</pre></div>
</div>
</section>
<section id="output">
<h1>Output<a class="headerlink" href="#output" title="Link to this heading">¶</a></h1>
<p>Above command produces</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">my_sorting_screen_filtered.h5ad</span></code> with filtered alleles stored in <code class="docutils literal notranslate"><span class="pre">.uns</span></code>,</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">my_sorting_screen_filtered.filtered_allele_stats.pdf</span></code>, and <code class="docutils literal notranslate"><span class="pre">my_sorting_screen_filtered.filter_log.txt</span></code> that report allele count stats in each filtering step.</p></li>
</ul>
<p>You may want to adjust the flitering parameters to obtain optimal balance between # guides per variant &amp; # variants that are scored. See example outputs of filtering step <a class="reference external" href="docs/example_filtering_output/">here</a>.</p>
</section>
<section id="translating-alleles">
<h1>Translating alleles<a class="headerlink" href="#translating-alleles" title="Link to this heading">¶</a></h1>
<p>If you want to obtain <strong>amino acid level variant</strong> for coding sequence tiling screens, provide coding sequence positions which variants occuring within the coding sequence will be translated. <em>This is optional, but **highly recommended*</em> to increase per-(coding)variant support.*</p>
<p><span class="raw-html-m2r"><img src="../../imgs/translation.png" alt="Allele translation" width="500"/></span></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>bean<span class="w"> </span>filter<span class="w"> </span>my_sorting_screen.h5ad<span class="w"> </span><span class="se">\</span>
-o<span class="w"> </span>my_sorting_screen_masked.h5ad<span class="w"> </span><span class="se">\</span>
--translate<span class="w">   </span><span class="sb">`</span><span class="c1"># Translate coding variants` \</span>
<span class="o">[</span><span class="w"> </span>--translate-gene-name<span class="w"> </span>GENE_SYMBOL<span class="w"> </span>OR
<span class="w">  </span>--translate-genes-list<span class="w"> </span>path_to_gene_names_file.txt<span class="w"> </span>OR
<span class="w">  </span>--translate-fasta<span class="w"> </span>gene_exon.fa,<span class="w"> </span>OR
<span class="w">  </span>--translate-fastas-csv<span class="w"> </span>gene_exon_fas.csv<span class="o">]</span>
</pre></div>
</div>
<ul class="simple">
<li><p>When library covers a single gene, do either of the following:</p>
<ol class="arabic simple">
<li><p>Feed <code class="docutils literal notranslate"><span class="pre">--translate-gene-name</span> <span class="pre">GENE_SYMBOL</span></code> if your <code class="docutils literal notranslate"><span class="pre">genomic_pos</span></code> column of <code class="docutils literal notranslate"><span class="pre">sgRNA_info_tbl</span></code> is compatible with <a class="reference external" href="https://useast.ensembl.org/info/genome/genebuild/mane.html">MANE transcript</a>‘s reference genome. (Per 10/23/2023, GRCh38). This will automatically load the exon positions based on MANE transcript annotation.</p></li>
<li><p>To use your custom coding sequence and exon positions, feed <code class="docutils literal notranslate"><span class="pre">--translate-fasta</span> <span class="pre">gene_exon.fa</span></code> argument where <code class="docutils literal notranslate"><span class="pre">gene_exon.fa</span></code> is the FASTA file with entries of exons. <a class="reference external" href="docs/exon_fa_format.md">See full details here</a>.</p></li>
</ol>
</li>
<li><p>When library covers multiple genes, do either of the following:</p>
<ol class="arabic simple">
<li><p>Feed <code class="docutils literal notranslate"><span class="pre">--translate-genes-list</span> <span class="pre">path_to_gene_names_file.txt</span></code> where <code class="docutils literal notranslate"><span class="pre">path_to_gene_names_file.txt</span></code> is file with one gene symbol per line.</p></li>
<li><p>Feed <code class="docutils literal notranslate"><span class="pre">--translate-fastas-csv</span> <span class="pre">gene_exon_fas.csv</span></code> where <code class="docutils literal notranslate"><span class="pre">gene_exon_fas.csv</span></code> is the csv file with lines <code class="docutils literal notranslate"><span class="pre">gene_id,gene_exon_fasta_path</span></code> without header. Each FASTA file in <code class="docutils literal notranslate"><span class="pre">gene_exon_fasta_path</span></code> is formatted <a class="reference external" href="docs/exon_fa_format.md">as the single-gene FASTA file</a>.</p></li>
</ol>
</li>
<li><p>Translation will keep the variants outside the coding sequence as nucleotide-level variants, while aggregating variants leading to the same coding sequence variants.</p></li>
</ul>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index_.html">bean</a></h1>








<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../gwas.html">GWAS variant library</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cds.html">Coding sequence tiling library</a></li>
<li class="toctree-l1"><a class="reference internal" href="../input.html">Input file format</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../subcommands.html">Subcommands</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../ReporterScreen_api.html">ReporterScreen API tutorial</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index_.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2024, Jayoung Ryu.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 7.2.6</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 0.7.16</a>
      
      |
      <a href="../_sources/commands/filter.md.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>