<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>sgRNA_info_table.csv &#8212; bean 1.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=4f649999" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css?v=039e1c02" />
    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js?v=af2ce170"></script>
    <script src="../_static/doctools.js?v=888ff710"></script>
    <script src="../_static/sphinx_highlight.js?v=4825356b"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <p>This document describes the input files of <a class="reference internal" href="../count_samples.html#count-samples"><span class="std std-ref">bean count-samples</span></a>.</p>
<section id="sgrna-info-table-csv">
<h1>sgRNA_info_table.csv<a class="headerlink" href="#sgrna-info-table-csv" title="Permalink to this heading">¶</a></h1>
<p>File should contain following columns.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">name</span></code>: gRNA ID column</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sequence</span></code>: gRNA sequence</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">barcode</span></code>: R2 barcode to help match reporter to gRNA, written in the sense direction (as in R1)</p></li>
<li><p>In order to use accessibility in the <a class="reference external" href="#bean-run-quantify-variant-effects">variant effect quantification</a>, provide accessibility information in one of two options. (For non-targeting guides, provide NA values (empty cell).)</p>
<ul>
<li><p>Option 1: <code class="docutils literal notranslate"><span class="pre">chrom</span></code> &amp; <code class="docutils literal notranslate"><span class="pre">genomic_pos</span></code>: Chromosome (ex. <code class="docutils literal notranslate"><span class="pre">chr19</span></code>) and genomic position of guide sequence. You will have to provide the path to the bigwig file with matching reference version in <code class="docutils literal notranslate"><span class="pre">bean</span> <span class="pre">run</span></code>.</p></li>
<li><p>Option 2: <code class="docutils literal notranslate"><span class="pre">accessibility_signal</span></code>: ATAC-seq signal value of the target loci of each guide.</p></li>
</ul>
</li>
<li><p>For variant library (gRNAs are designed to target specific variants and ignores bystander edits)</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">target</span></code>: This column denotes which target variant/element of each gRNA. This is not used in <code class="docutils literal notranslate"><span class="pre">bean</span> <span class="pre">count[-samples]</span></code> but required to run <code class="docutils literal notranslate"><span class="pre">bean</span> <span class="pre">run</span></code> in later steps.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">target_group</span></code>: If negative/positive control gRNA will be considered in <code class="docutils literal notranslate"><span class="pre">bean</span> <span class="pre">qc</span></code> and/or <code class="docutils literal notranslate"><span class="pre">bean</span> <span class="pre">run</span></code>, specify as “NegCtrl”/”PosCtrl” in this column.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">target_pos</span></code>: If <code class="docutils literal notranslate"><span class="pre">--match_target_pos</span></code> flag is used, input file needs <code class="docutils literal notranslate"><span class="pre">target_pos</span></code> which specifies 0-based relative position of targeted base within Reporter sequence.</p></li>
</ul>
</li>
<li><p>For tiling library (gRNAs tile coding / noncoding sequences)</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">strand</span></code>: Specifies gRNA strand information relative to the reference genome.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">chrom</span></code>: Chromosome of gRNA targeted locus.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">start_pos</span></code>: gRNA starting position in the genome. Required when you provide <code class="docutils literal notranslate"><span class="pre">strand</span></code> column. Should specify the smaller coordinate value among start and end position regardless of gRNA strandedness.</p></li>
</ul>
</li>
</ul>
<p>Also see examples for <a class="reference external" href="tests/data/test_guide_info.csv">variant library</a> and <a class="reference external" href="tests/data/test_guide_info_tiling.csv">tiling library</a>.</p>
</section>
<section id="sample-list-csv">
<h1>sample_list.csv<a class="headerlink" href="#sample-list-csv" title="Permalink to this heading">¶</a></h1>
<p>File should contain following columns with header.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">R1_filepath</span></code>: Path to read 1 <code class="docutils literal notranslate"><span class="pre">.fastq[.gz]</span></code> file</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">R2_filepath</span></code>: Path to read 1 <code class="docutils literal notranslate"><span class="pre">.fastq[.gz]</span></code> file</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sample_id</span></code>: ID of sequencing sample</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">replicate</span></code>: Replicate # of this sample (Should NOT contain <code class="docutils literal notranslate"><span class="pre">.</span></code>)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">condition</span></code>: Name of the sorting bin (ex. <code class="docutils literal notranslate"><span class="pre">top</span></code>, <code class="docutils literal notranslate"><span class="pre">bot</span></code>), or label of timepoint (ex. <code class="docutils literal notranslate"><span class="pre">D5</span></code>, <code class="docutils literal notranslate"><span class="pre">D18</span></code>)</p></li>
</ul>
<p>For FACS sorting screens:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">upper_quantile</span></code>: FACS sorting upper quantile</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">lower_quantile</span></code>: FACS sorting lower quantile</p></li>
</ul>
<p>For proliferation / survival screens:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">time</span></code>: Numeric time following the base editing of each sample.</p></li>
</ul>
<p>Also see examples for <a class="reference external" href="tests/data/sample_list.csv">FACS sorting screen</a>.</p>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">bean</a></h1>








<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../gwas.html">GWAS variant library</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cds.html">Coding sequence tiling library</a></li>
<li class="toctree-l1"><a class="reference internal" href="../input.html">Input file format</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../subcommands.html">Subcommands</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../ReporterScreen_api.html">ReporterScreen API tutorial</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2024, Jayoung Ryu.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 7.1.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.13</a>
      
      |
      <a href="../_sources/commands/input.md.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>