<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>bean count &#8212; bean 1.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=d1102ebc" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=12dfc556" />
    <script src="_static/documentation_options.js?v=8d563738"></script>
    <script src="_static/doctools.js?v=888ff710"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="bean count-samples" href="count_samples.html" />
    <link rel="prev" title="Subcommands" href="subcommands.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="bean-count">
<h1><cite>bean count</cite><a class="headerlink" href="#bean-count" title="Link to this heading">¶</a></h1>
<section id="bean-count-samples-count-reporter-screen-data">
<h2><code class="docutils literal notranslate"><span class="pre">bean</span> <span class="pre">count[-samples]</span></code>: Count (reporter) screen data<a class="headerlink" href="#bean-count-samples-count-reporter-screen-data" title="Link to this heading">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">bean</span> <span class="pre">count-samples</span></code> (or <code class="docutils literal notranslate"><span class="pre">bean</span> <span class="pre">count</span></code> for a single sample) maps guide into guide counts, <strong>allowing for base transition in spacer sequence</strong>. When the matched reporter information is provided, it can count the <strong>target site edits</strong> and <strong>alleles produced by each guide</strong>. Mapping is efficiently done based on <a class="reference external" href="https://github.com/pinellolab/CRISPResso2">CRISPResso2</a> modified for base-edit-aware mapping.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">bean</span> <span class="n">count</span><span class="o">-</span><span class="n">samples</span> \
  <span class="o">--</span><span class="nb">input</span> <span class="n">sample_list</span><span class="o">.</span><span class="n">csv</span>   <span class="err">`</span><span class="c1"># sample with lines &#39;R1_filepath,R2_filepath,sample_name\n&#39;` \</span>
  <span class="o">-</span><span class="n">b</span> <span class="n">A</span>                      <span class="err">`</span><span class="c1"># base that is being edited (A/G)` \</span>
  <span class="o">-</span><span class="n">f</span> <span class="n">sgRNA_info_table</span><span class="o">.</span><span class="n">csv</span>   <span class="err">`</span><span class="c1"># sgRNA information` \</span>
  <span class="o">-</span><span class="n">o</span> <span class="o">.</span>                      <span class="err">`</span><span class="c1"># output directory` \</span>
  <span class="o">-</span><span class="n">r</span>                        <span class="err">`</span><span class="c1"># read edit/allele information from reporter` \</span>
  <span class="o">-</span><span class="n">t</span> <span class="mi">12</span>                     <span class="err">`</span><span class="c1"># number of threads` \</span>
  <span class="o">--</span><span class="n">name</span> <span class="n">my_sorting_screen</span>  <span class="err">`</span><span class="c1"># name of this sample run` \</span>
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">bean</span> <span class="n">count</span> <span class="o">--</span><span class="n">R1</span> <span class="n">R1</span><span class="o">.</span><span class="n">fq</span> <span class="o">--</span><span class="n">R2</span> <span class="n">R2</span><span class="o">.</span><span class="n">fq</span> <span class="o">-</span><span class="n">b</span> <span class="n">A</span> <span class="o">-</span><span class="n">f</span> <span class="n">sgRNA_info_table</span><span class="o">.</span><span class="n">csv</span> <span class="o">-</span><span class="n">r</span>
</pre></div>
</div>
<p>By default, <code class="docutils literal notranslate"><span class="pre">bean</span> <span class="pre">count[-samples]</span></code> assume R1 and R2 are trimmed off of the adapter sequence. You may need to adjust the command arguments according to your read structure.</p>
<blockquote>
<div><p><span class="raw-html-m2r"><img src="../../imgs/sequence_struct.png" alt="Read structuren" width="600"/></span></p>
</div></blockquote>
<p>See full detail <a class="reference external" href="#full-parameters">below</a>.</p>
</section>
<section id="input-file-format">
<h2>Input file format<a class="headerlink" href="#input-file-format" title="Link to this heading">¶</a></h2>
<p>See <a class="reference internal" href="input.html#input"><span class="std std-ref">Input file format</span></a> for input file formats.</p>
</section>
<section id="output-file-format">
<h2>Output file format<a class="headerlink" href="#output-file-format" title="Link to this heading">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">count</span></code> or <code class="docutils literal notranslate"><span class="pre">count-samples</span></code> produces <code class="docutils literal notranslate"><span class="pre">.h5ad</span></code> and <code class="docutils literal notranslate"><span class="pre">.xlsx</span></code> file with guide and per-guide allele counts.</p>
<ul class="simple">
<li><p><cite>.h5ad</cite>: This output file follows annotated matrix format compatible with <cite>AnnData</cite> and is based on <cite>Screen</cite> object in [purturb_tools](<a class="reference external" href="https://github.com/pinellolab/perturb-tools">https://github.com/pinellolab/perturb-tools</a>). See <a class="reference external" href="#data-structure">Data Structure</a> section for more information.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">.xlsx</span></code>: This output file contains <code class="docutils literal notranslate"><span class="pre">.guides</span></code>, <code class="docutils literal notranslate"><span class="pre">.samples</span></code>, <code class="docutils literal notranslate"><span class="pre">.X[_bcmatch,_edits]</span></code>. (<code class="docutils literal notranslate"><span class="pre">allele_tables</span></code> are often too large to write into an Excel!)</p></li>
</ul>
</section>
<section id="full-parameters">
<h2>Full parameters<a class="headerlink" href="#full-parameters" title="Link to this heading">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">bean</span> <span class="n">count</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="o">--</span><span class="n">R1</span> <span class="n">R1</span> <span class="o">--</span><span class="n">R2</span> <span class="n">R2</span> <span class="o">-</span><span class="n">b</span> <span class="n">EDITED_BASE</span> <span class="o">-</span><span class="n">f</span> <span class="n">SGRNA_FILENAME</span>
                  <span class="p">[</span><span class="o">--</span><span class="n">guide</span><span class="o">-</span><span class="n">start</span><span class="o">-</span><span class="n">seq</span> <span class="n">GUIDE_START_SEQ</span><span class="p">]</span>
                  <span class="p">[</span><span class="o">--</span><span class="n">guide</span><span class="o">-</span><span class="n">end</span><span class="o">-</span><span class="n">seq</span> <span class="n">GUIDE_END_SEQ</span><span class="p">]</span>
                  <span class="p">[</span><span class="o">--</span><span class="n">barcode</span><span class="o">-</span><span class="n">start</span><span class="o">-</span><span class="n">seq</span> <span class="n">BARCODE_START_SEQ</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">r</span><span class="p">]</span>
                  <span class="p">[</span><span class="o">-</span><span class="n">q</span> <span class="n">MIN_AVERAGE_READ_QUALITY</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">MIN_SINGLE_BP_QUALITY</span><span class="p">]</span>
                  <span class="p">[</span><span class="o">-</span><span class="n">n</span> <span class="n">NAME</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT_FOLDER</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">REPORTER_LENGTH</span><span class="p">]</span>
                  <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">intermediate</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">qstart</span><span class="o">-</span><span class="n">R1</span> <span class="n">QSTART_R1</span><span class="p">]</span>
                  <span class="p">[</span><span class="o">--</span><span class="n">qend</span><span class="o">-</span><span class="n">R1</span> <span class="n">QEND_R1</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">qstart</span><span class="o">-</span><span class="n">R2</span> <span class="n">QSTART_R2</span><span class="p">]</span>
                  <span class="p">[</span><span class="o">--</span><span class="n">qend</span><span class="o">-</span><span class="n">R2</span> <span class="n">QEND_R2</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">gstart</span><span class="o">-</span><span class="n">reporter</span> <span class="n">GSTART_REPORTER</span><span class="p">]</span>
                  <span class="p">[</span><span class="o">--</span><span class="n">match</span><span class="o">-</span><span class="n">target</span><span class="o">-</span><span class="n">pos</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">target</span><span class="o">-</span><span class="n">pos</span><span class="o">-</span><span class="n">col</span> <span class="n">TARGET_POS_COL</span><span class="p">]</span>
                  <span class="p">[</span><span class="o">--</span><span class="n">guide</span><span class="o">-</span><span class="n">bc</span> <span class="n">GUIDE_BC</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">guide</span><span class="o">-</span><span class="n">bc</span><span class="o">-</span><span class="nb">len</span> <span class="n">GUIDE_BC_LEN</span><span class="p">]</span>
                  <span class="p">[</span><span class="o">--</span><span class="n">offset</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">align</span><span class="o">-</span><span class="n">fasta</span> <span class="n">ALIGN_FASTA</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">string</span><span class="o">-</span><span class="n">allele</span><span class="p">]</span>
                  <span class="p">[</span><span class="o">-</span><span class="n">g</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">m</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">tiling</span><span class="p">]</span>
</pre></div>
</div>
<section id="named-arguments">
<h3>Named Arguments<a class="headerlink" href="#named-arguments" title="Link to this heading">¶</a></h3>
<dl class="option-list">
<dt><kbd>--R1</kbd></dt>
<dd><p>FASTQ file path for read 1</p>
</dd>
<dt><kbd>--R2</kbd></dt>
<dd><p>FASTQ file path for read 2.</p>
</dd>
<dt><kbd>-b, --edited-base</kbd></dt>
<dd><p>For base editors, the base that should be ignored when matching the gRNA sequence</p>
</dd>
<dt><kbd>-f, --sgRNA-filename</kbd></dt>
<dd><p>sgRNA description file. The format requires three columns: name, sequence, barcode [ reporter [,strand, target_pos], [start_pos, offset] ].</p>
</dd>
<dt><kbd>--guide-start-seq</kbd></dt>
<dd><p>Guide starts after this sequence in R1</p>
<p>Default: “”</p>
</dd>
<dt><kbd>--guide-end-seq</kbd></dt>
<dd><p>Guide starts after this sequence in R1</p>
<p>Default: “”</p>
</dd>
<dt><kbd>--barcode-start-seq</kbd></dt>
<dd><p>Barcode + reporter starts after this sequence in R2, denoted as the sense direction (the same sequence direction as R1).</p>
<p>Default: “”</p>
</dd>
<dt><kbd>-r, --count-reporter</kbd></dt>
<dd><p>Count reporter edits.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-q, --min-average-read-quality</kbd></dt>
<dd><p>Minimum average quality score (phred33) to keep a read</p>
<p>Default: 30</p>
</dd>
<dt><kbd>-s, --min-single-bp-quality</kbd></dt>
<dd><p>Minimum single bp score (phred33) to keep a read</p>
<p>Default: 0</p>
</dd>
<dt><kbd>-n, --name</kbd></dt>
<dd><p>Output name</p>
<p>Default: “”</p>
</dd>
<dt><kbd>-o, --output-folder</kbd></dt>
<dd><p>Default: “”</p>
</dd>
<dt><kbd>-l, --reporter-length</kbd></dt>
<dd><p>length of the reporter</p>
<p>Default: 32</p>
</dd>
<dt><kbd>--keep-intermediate</kbd></dt>
<dd><p>Keep all the  intermediate files</p>
<p>Default: False</p>
</dd>
<dt><kbd>--qstart-R1</kbd></dt>
<dd><p>Start position of the read when filtering for quality score of the read 1</p>
<p>Default: 0</p>
</dd>
<dt><kbd>--qend-R1</kbd></dt>
<dd><p>End position of the read when filtering for quality score of the read 1</p>
<p>Default: 47</p>
</dd>
<dt><kbd>--qstart-R2</kbd></dt>
<dd><p>Same as qstart_R1, for read 2 fastq file</p>
<p>Default: 0</p>
</dd>
<dt><kbd>--qend-R2</kbd></dt>
<dd><p>Same as qstart_R2, for read 2 fastq file</p>
<p>Default: 36</p>
</dd>
<dt><kbd>--gstart-reporter</kbd></dt>
<dd><p>Start position of the guide sequence in the reporter</p>
<p>Default: 6</p>
</dd>
<dt><kbd>--match-target-pos</kbd></dt>
<dd><p>Count the edit in the exact target position.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--target-pos-col</kbd></dt>
<dd><p>Column name specifying the relative target position within <em>reporter</em> sequence.</p>
<p>Default: “target_pos”</p>
</dd>
<dt><kbd>--guide-bc</kbd></dt>
<dd><p>Construct has guide barcode</p>
<p>Default: True</p>
</dd>
<dt><kbd>--guide-bc-len</kbd></dt>
<dd><p>Guide barcode sequence length at the beginning of the R2</p>
<p>Default: 4</p>
</dd>
<dt><kbd>--offset</kbd></dt>
<dd><p>Guide file has <cite>offest</cite> column that will be added to the relative position of reporters.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--align-fasta</kbd></dt>
<dd><p>gRNA is aligned to this sequence to infer the offset. Can be used when the exact offset is not provided.</p>
<p>Default: “”</p>
</dd>
<dt><kbd>--string-allele</kbd></dt>
<dd><p>Store allele as quality filtered string instead of Allele object</p>
<p>Default: False</p>
</dd>
<dt><kbd>-g, --count-guide-edits</kbd></dt>
<dd><p>count the self editing of guides</p>
<p>Default: False</p>
</dd>
<dt><kbd>-m, --count-guide-reporter-alleles</kbd></dt>
<dd><p>count the matched allele of guide and reporter edit</p>
<p>Default: False</p>
</dd>
<dt><kbd>--tiling</kbd></dt>
<dd><p>Specify that the guide library is tiling library without ‘n guides per target’ design</p>
<p>Default: False</p>
</dd>
</dl>
</section>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index_.html">bean</a></h1>








<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="gwas.html">GWAS variant library</a></li>
<li class="toctree-l1"><a class="reference internal" href="cds.html">Coding sequence tiling library</a></li>
<li class="toctree-l1"><a class="reference internal" href="input.html">Input file format</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="subcommands.html">Subcommands</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#"><cite>bean count</cite></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#bean-count-samples-count-reporter-screen-data"><code class="docutils literal notranslate"><span class="pre">bean</span> <span class="pre">count[-samples]</span></code>: Count (reporter) screen data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#input-file-format">Input file format</a></li>
<li class="toctree-l3"><a class="reference internal" href="#output-file-format">Output file format</a></li>
<li class="toctree-l3"><a class="reference internal" href="#full-parameters">Full parameters</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="count_samples.html"><cite>bean count-samples</cite></a></li>
<li class="toctree-l2"><a class="reference internal" href="profile.html"><cite>bean profile</cite></a></li>
<li class="toctree-l2"><a class="reference internal" href="qc.html"><cite>bean qc</cite></a></li>
<li class="toctree-l2"><a class="reference internal" href="filter.html"><cite>bean filter</cite></a></li>
<li class="toctree-l2"><a class="reference internal" href="run.html"><cite>bean run</cite></a></li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="ReporterScreen_api.html">ReporterScreen API tutorial</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index_.html">Documentation overview</a><ul>
  <li><a href="subcommands.html">Subcommands</a><ul>
      <li>Previous: <a href="subcommands.html" title="previous chapter">Subcommands</a></li>
      <li>Next: <a href="count_samples.html" title="next chapter"><cite>bean count-samples</cite></a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2024, Jayoung Ryu.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 7.2.6</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 0.7.16</a>
      
      |
      <a href="_sources/count.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>