<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>bean filter &#8212; bean 1.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=4f649999" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=039e1c02" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js?v=af2ce170"></script>
    <script src="_static/doctools.js?v=888ff710"></script>
    <script src="_static/sphinx_highlight.js?v=4825356b"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="bean run" href="run.html" />
    <link rel="prev" title="bean qc" href="qc.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="bean-filter">
<span id="filter"></span><h1><cite>bean filter</cite><a class="headerlink" href="#bean-filter" title="Permalink to this heading">¶</a></h1>
<section id="filter-filtering-and-optionally-translating-alleles">
<h2><code class="docutils literal notranslate"><span class="pre">filter</span></code>: Filtering (and optionally translating) alleles<a class="headerlink" href="#filter-filtering-and-optionally-translating-alleles" title="Permalink to this heading">¶</a></h2>
<p>As <code class="docutils literal notranslate"><span class="pre">tiling</span></code> mode of <code class="docutils literal notranslate"><span class="pre">bean</span> <span class="pre">run</span></code> accounts for any robustly observed alleles, <code class="docutils literal notranslate"><span class="pre">bean</span> <span class="pre">filter</span></code> filters for such alleles.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>bean<span class="w"> </span>filter<span class="w"> </span>my_sorting_screen_masked.h5ad<span class="w"> </span><span class="se">\</span>
-o<span class="w"> </span>my_sorting_screen_filtered.h5ad<span class="w">  </span><span class="sb">`</span><span class="c1"># Output file path` \</span>
</pre></div>
</div>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this heading">¶</a></h2>
<p>Above command produces</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">my_sorting_screen_filtered.h5ad</span></code> with filtered alleles stored in <code class="docutils literal notranslate"><span class="pre">.uns</span></code>,</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">my_sorting_screen_filtered.filtered_allele_stats.pdf</span></code>, and <code class="docutils literal notranslate"><span class="pre">my_sorting_screen_filtered.filter_log.txt</span></code> that report allele count stats in each filtering step.</p></li>
</ul>
<p>You may want to adjust the flitering parameters to obtain optimal balance between # guides per variant &amp; # variants that are scored. See example outputs of filtering step <a class="reference external" href="docs/example_filtering_output/">here</a>.</p>
</section>
<section id="translating-alleles">
<h2>Translating alleles<a class="headerlink" href="#translating-alleles" title="Permalink to this heading">¶</a></h2>
<p>If you want to obtain <strong>amino acid level variant</strong> for coding sequence tiling screens, provide coding sequence positions which variants occuring within the coding sequence will be translated. <em>This is optional, but **highly recommended*</em> to increase per-(coding)variant support.*</p>
<p><span class="raw-html-m2r"><img src="../../imgs/translation.png" alt="Allele translation" width="500"/></span></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>bean<span class="w"> </span>filter<span class="w"> </span>my_sorting_screen.h5ad<span class="w"> </span><span class="se">\</span>
-o<span class="w"> </span>my_sorting_screen_masked.h5ad<span class="w"> </span><span class="se">\</span>
--translate<span class="w">   </span><span class="sb">`</span><span class="c1"># Translate coding variants` \</span>
<span class="o">[</span><span class="w"> </span>--translate-gene-name<span class="w"> </span>GENE_SYMBOL<span class="w"> </span>OR
<span class="w">  </span>--translate-genes-list<span class="w"> </span>path_to_gene_names_file.txt<span class="w"> </span>OR
<span class="w">  </span>--translate-fasta<span class="w"> </span>gene_exon.fa,<span class="w"> </span>OR
<span class="w">  </span>--translate-fastas-csv<span class="w"> </span>gene_exon_fas.csv<span class="o">]</span>
</pre></div>
</div>
<ul class="simple">
<li><p>When library covers a single gene, do either of the following:</p>
<ol class="arabic simple">
<li><p>Feed <code class="docutils literal notranslate"><span class="pre">--translate-gene-name</span> <span class="pre">GENE_SYMBOL</span></code> if your <code class="docutils literal notranslate"><span class="pre">genomic_pos</span></code> column of <code class="docutils literal notranslate"><span class="pre">sgRNA_info_tbl</span></code> is compatible with <a class="reference external" href="https://useast.ensembl.org/info/genome/genebuild/mane.html">MANE transcript</a>‘s reference genome. (Per 10/23/2023, GRCh38). This will automatically load the exon positions based on MANE transcript annotation.</p></li>
<li><p>To use your custom coding sequence and exon positions, feed <code class="docutils literal notranslate"><span class="pre">--translate-fasta</span> <span class="pre">gene_exon.fa</span></code> argument where <code class="docutils literal notranslate"><span class="pre">gene_exon.fa</span></code> is the FASTA file with entries of exons. <a class="reference external" href="docs/exon_fa_format.md">See full details here</a>.</p></li>
</ol>
</li>
<li><p>When library covers multiple genes, do either of the following:</p>
<ol class="arabic simple">
<li><p>Feed <code class="docutils literal notranslate"><span class="pre">--translate-genes-list</span> <span class="pre">path_to_gene_names_file.txt</span></code> where <code class="docutils literal notranslate"><span class="pre">path_to_gene_names_file.txt</span></code> is file with one gene symbol per line.</p></li>
<li><p>Feed <code class="docutils literal notranslate"><span class="pre">--translate-fastas-csv</span> <span class="pre">gene_exon_fas.csv</span></code> where <code class="docutils literal notranslate"><span class="pre">gene_exon_fas.csv</span></code> is the csv file with lines <code class="docutils literal notranslate"><span class="pre">gene_id,gene_exon_fasta_path</span></code> without header. Each FASTA file in <code class="docutils literal notranslate"><span class="pre">gene_exon_fasta_path</span></code> is formatted <a class="reference external" href="docs/exon_fa_format.md">as the single-gene FASTA file</a>.</p></li>
</ol>
</li>
<li><p>Translation will keep the variants outside the coding sequence as nucleotide-level variants, while aggregating variants leading to the same coding sequence variants.</p></li>
</ul>
</section>
<section id="full-parameters">
<h2>Full parameters<a class="headerlink" href="#full-parameters" title="Permalink to this heading">¶</a></h2>
<p><p>Filter alleles based on edit position in spacer and frequency across samples.</p>
</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">bean</span> <span class="nb">filter</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">prefix</span> <span class="n">OUTPUT_PREFIX</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">plasmid</span><span class="o">-</span><span class="n">path</span> <span class="n">PLASMID_PATH</span><span class="p">]</span>
                   <span class="p">[</span><span class="o">--</span><span class="n">edit</span><span class="o">-</span><span class="n">start</span><span class="o">-</span><span class="n">pos</span> <span class="n">EDIT_START_POS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">edit</span><span class="o">-</span><span class="n">end</span><span class="o">-</span><span class="n">pos</span> <span class="n">EDIT_END_POS</span><span class="p">]</span>
                   <span class="p">[</span><span class="o">--</span><span class="n">jaccard</span><span class="o">-</span><span class="n">threshold</span> <span class="n">JACCARD_THRESHOLD</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="nb">filter</span><span class="o">-</span><span class="n">window</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">indels</span><span class="p">]</span>
                   <span class="p">[</span><span class="o">--</span><span class="nb">filter</span><span class="o">-</span><span class="n">target</span><span class="o">-</span><span class="n">basechange</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">translate</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">translate</span><span class="o">-</span><span class="n">fasta</span> <span class="n">TRANSLATE_FASTA</span><span class="p">]</span>
                   <span class="p">[</span><span class="o">--</span><span class="n">translate</span><span class="o">-</span><span class="n">fastas</span><span class="o">-</span><span class="n">csv</span> <span class="n">TRANSLATE_FASTAS_CSV</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">translate</span><span class="o">-</span><span class="n">gene</span> <span class="n">TRANSLATE_GENE</span><span class="p">]</span>
                   <span class="p">[</span><span class="o">--</span><span class="n">translate</span><span class="o">-</span><span class="n">genes</span><span class="o">-</span><span class="nb">list</span> <span class="n">TRANSLATE_GENES_LIST</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="nb">filter</span><span class="o">-</span><span class="n">allele</span><span class="o">-</span><span class="n">proportion</span> <span class="n">FILTER_ALLELE_PROPORTION</span><span class="p">]</span>
                   <span class="p">[</span><span class="o">--</span><span class="nb">filter</span><span class="o">-</span><span class="n">allele</span><span class="o">-</span><span class="n">count</span> <span class="n">FILTER_ALLELE_COUNT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="nb">filter</span><span class="o">-</span><span class="n">sample</span><span class="o">-</span><span class="n">proportion</span> <span class="n">FILTER_SAMPLE_PROPORTION</span><span class="p">]</span>
                   <span class="p">[</span><span class="o">--</span><span class="n">load</span><span class="o">-</span><span class="n">tmp</span><span class="p">]</span>
                   <span class="n">bdata_path</span>
</pre></div>
</div>
<section id="positional-arguments">
<h3>Positional Arguments<a class="headerlink" href="#positional-arguments" title="Permalink to this heading">¶</a></h3>
<dl class="option-list">
<dt><kbd>bdata_path</kbd></dt>
<dd><p>Input ReporterScreen file of which allele will be filtered out.</p>
</dd>
</dl>
</section>
<section id="named-arguments">
<h3>Named Arguments<a class="headerlink" href="#named-arguments" title="Permalink to this heading">¶</a></h3>
<dl class="option-list">
<dt><kbd>--output-prefix, -o</kbd></dt>
<dd><p>Output prefix for log and ReporterScreen file with allele assignment</p>
</dd>
<dt><kbd>--plasmid-path, -p</kbd></dt>
<dd><p>Plasmid ReporterScreen object path. If provided, alleles are filtered based on if a nucleotide edit is more significantly enriched in sample compared to the plasmid data. Negative control data where no edit is expected can be fed in instead of plasmid library.</p>
</dd>
<dt><kbd>--edit-start-pos, -s</kbd></dt>
<dd><p>0-based start posiiton (inclusive) of edit relative to the start of guide spacer.</p>
<p>Default: 2</p>
</dd>
<dt><kbd>--edit-end-pos, -e</kbd></dt>
<dd><p>0-based end position (exclusive) of edit relative to the start of guide spacer.</p>
<p>Default: 7</p>
</dd>
<dt><kbd>--jaccard-threshold, -j</kbd></dt>
<dd><p>Jaccard Index threshold when the alleles are mapped to the most similar alleles. In each filtering step, allele counts of filtered out alleles will be mapped to the most similar allele only if they have Jaccard Index of shared edit higher than this threshold.</p>
<p>Default: 0.3</p>
</dd>
<dt><kbd>--filter-window, -w</kbd></dt>
<dd><p>Only consider edit within window provided by (edit-start-pos, edit-end-pos). If this flag is not provided, <cite>–edit-start-pos</cite> and <cite>–edit-end-pos</cite> flags are ignored.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--keep-indels, -i</kbd></dt>
<dd><p>Include indels.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--filter-target-basechange, -b</kbd></dt>
<dd><p>Only consider target edit (stored in bdata.uns[‘target_base_change’])</p>
<p>Default: False</p>
</dd>
<dt><kbd>--translate, -t</kbd></dt>
<dd><p>Translate alleles</p>
<p>Default: False</p>
</dd>
<dt><kbd>--translate-fasta, -f</kbd></dt>
<dd><p>fasta file path with exon positions. If not provided, LDLR hg19 coordinates will be used.</p>
</dd>
<dt><kbd>--translate-fastas-csv, -fs</kbd></dt>
<dd><p>.csv with two columns with gene IDs and FASTA file path corresponding to each gene.</p>
</dd>
<dt><kbd>--translate-gene, -g</kbd></dt>
<dd><p>Gene symbol if a gene is tiled. If not provided, LDLR hg19 coordinates will be used.</p>
</dd>
<dt><kbd>--translate-genes-list, -gs</kbd></dt>
<dd><p>File with gene symbols, one per line, if multiple genes are tiled.</p>
</dd>
<dt><kbd>--filter-allele-proportion, -ap</kbd></dt>
<dd><p>If provided, alleles that exceed <cite>filter_allele_proportion</cite> in <cite>filter-sample-proportion</cite> will be retained.</p>
<p>Default: 0.05</p>
</dd>
<dt><kbd>--filter-allele-count, -ac</kbd></dt>
<dd><p>If provided, alleles that exceed <cite>filter_allele_proportion</cite> AND <cite>filter_allele_count</cite> in <cite>filter-sample-proportion</cite> will be retained.</p>
<p>Default: 5</p>
</dd>
<dt><kbd>--filter-sample-proportion, -sp</kbd></dt>
<dd><p>If <cite>filter_allele_proportion</cite> is provided, alleles that exceed <cite>filter_allele_proportion</cite> in <cite>filter-sample-proportion</cite> will be retained.</p>
<p>Default: 0.2</p>
</dd>
<dt><kbd>--load-tmp</kbd></dt>
<dd><p>Load temporary file and work from there.</p>
<p>Default: False</p>
</dd>
</dl>
</section>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">bean</a></h1>








<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="gwas.html">GWAS variant library</a></li>
<li class="toctree-l1"><a class="reference internal" href="cds.html">Coding sequence tiling library</a></li>
<li class="toctree-l1"><a class="reference internal" href="input.html">Input file format</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="subcommands.html">Subcommands</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="count.html"><cite>bean count</cite></a></li>
<li class="toctree-l2"><a class="reference internal" href="count_samples.html"><cite>bean count-samples</cite></a></li>
<li class="toctree-l2"><a class="reference internal" href="profile.html"><cite>bean profile</cite></a></li>
<li class="toctree-l2"><a class="reference internal" href="qc.html"><cite>bean qc</cite></a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#"><cite>bean filter</cite></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#filter-filtering-and-optionally-translating-alleles"><code class="docutils literal notranslate"><span class="pre">filter</span></code>: Filtering (and optionally translating) alleles</a></li>
<li class="toctree-l3"><a class="reference internal" href="#output">Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#translating-alleles">Translating alleles</a></li>
<li class="toctree-l3"><a class="reference internal" href="#full-parameters">Full parameters</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="run.html"><cite>bean run</cite></a></li>
<li class="toctree-l2"><a class="reference internal" href="create_screen.html"><cite>bean create-screen</cite></a></li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="ReporterScreen_api.html">ReporterScreen API tutorial</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="subcommands.html">Subcommands</a><ul>
      <li>Previous: <a href="qc.html" title="previous chapter"><cite>bean qc</cite></a></li>
      <li>Next: <a href="run.html" title="next chapter"><cite>bean run</cite></a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2024, Jayoung Ryu.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 7.1.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.13</a>
      
      |
      <a href="_sources/filter.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>