<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>GWAS variant library &#8212; bean 1.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=d1102ebc" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=12dfc556" />
    <script src="_static/documentation_options.js?v=8d563738"></script>
    <script src="_static/doctools.js?v=888ff710"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Coding sequence tiling library" href="cds.html" />
    <link rel="prev" title="Welcome to bean’s documentation!" href="index_.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="gwas-variant-library">
<h1>GWAS variant library<a class="headerlink" href="#gwas-variant-library" title="Link to this heading">¶</a></h1>
<section id="variant-sorting-screen-tutorial">
<h2>Variant sorting screen tutorial<a class="headerlink" href="#variant-sorting-screen-tutorial" title="Link to this heading">¶</a></h2>
<p>GWAS variant screen with per-variant gRNA tiling design, selected based on FACS signal quantiles.</p>
<table>
  <tr>
    <th>Library design</th>
    <td>Variant (gRNAs tile each target variant)   <br> <img src="../../imgs/variant.png" alt="variant library design" width="600"/></td>
  </tr>
  <tr>
    <th>Selection</th>
    <td>Cells are sorted based on FACS signal quantiles  <br>  <img src="../../imgs/sorting_bins@8x.png" alt="variant library design" width="300"/></td>
  </tr>
</table><p><span class="raw-html-m2r"><br></br></span></p>
<section id="example-workflow">
<h3>Example workflow<a class="headerlink" href="#example-workflow" title="Link to this heading">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nv">screen_id</span><span class="o">=</span>my_sorting_tiling_screen

<span class="c1"># 1. Count gRNA &amp; reporter</span>
bean-count-samples<span class="w"> </span><span class="se">\</span>
--input<span class="w"> </span>tests/data/sample_list.csv<span class="w">    </span><span class="sb">`</span><span class="c1"># Contains fastq file path; see test file for example.`\</span>
-b<span class="w"> </span>A<span class="w">                                  </span><span class="sb">`</span><span class="c1"># Base A is edited (into G)` \</span>
-f<span class="w"> </span>tests/data/test_guide_info.csv<span class="w">     </span><span class="sb">`</span><span class="c1"># Contains gRNA metadata; see test file for example.`\</span>
-o<span class="w"> </span>./<span class="w">                                 </span><span class="sb">`</span><span class="c1"># Output directory` \</span>
-r<span class="w">                                    </span><span class="sb">`</span><span class="c1"># Quantify reporter edits` \</span>
-n<span class="w"> </span><span class="si">${</span><span class="nv">screen_id</span><span class="si">}</span><span class="w">                          </span><span class="sb">`</span><span class="c1"># ID of the screen to be counted`</span>

<span class="c1"># 2. QC samples &amp; guides</span>
bean-qc<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>bean_count_<span class="si">${</span><span class="nv">screen_id</span><span class="si">}</span>.h5ad<span class="w">             </span><span class="sb">`</span><span class="c1"># Input ReporterScreen .h5ad file path` \</span>
<span class="w">  </span>-o<span class="w"> </span>bean_count_<span class="si">${</span><span class="nv">screen_id</span><span class="si">}</span>_masked.h5ad<span class="w">   </span><span class="sb">`</span><span class="c1"># Output ReporterScreen .h5ad file path` \</span>
<span class="w">  </span>-r<span class="w"> </span>qc_report_<span class="si">${</span><span class="nv">screen_id</span><span class="si">}</span><span class="w">                </span><span class="sb">`</span><span class="c1"># Prefix for QC report` \</span>
<span class="w">  </span>-b<span class="w">                                       </span><span class="sb">`</span><span class="w"> </span><span class="c1"># Remove replicates with no good samples.</span>

<span class="c1"># 3. Quantify variant effect</span>
bean-run<span class="w"> </span>sorting<span class="w"> </span>variant<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>tests/data/bean_count_<span class="si">${</span><span class="nv">screen_id</span><span class="si">}</span>_masked.h5ad<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-o<span class="w"> </span>tests/test_res/var/<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--fit-negctrl<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--scale-by-acc<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--accessibility-col<span class="w"> </span>accessibility
</pre></div>
</div>
<p>See more details below.</p>
</section>
<section id="count-grna-reporter-count-samples">
<h3>1. Count gRNA &amp; reporter (<a class="reference internal" href="count_samples.html#count-samples"><span class="std std-ref">bean count-samples</span></a>)<a class="headerlink" href="#count-grna-reporter-count-samples" title="Link to this heading">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nv">screen_id</span><span class="o">=</span>my_sorting_tiling_screen

<span class="c1"># 1. Count gRNA &amp; reporter</span>
bean-count-samples<span class="w"> </span><span class="se">\</span>
--input<span class="w"> </span>tests/data/sample_list.csv<span class="w">    </span><span class="sb">`</span><span class="c1"># Contains fastq file path; see test file for example.`\</span>
-b<span class="w"> </span>A<span class="w">                                  </span><span class="sb">`</span><span class="c1"># Base A is edited (into G)` \</span>
-f<span class="w"> </span>tests/data/test_guide_info.csv<span class="w">     </span><span class="sb">`</span><span class="c1"># Contains gRNA metadata; see test file for example.`\</span>
-o<span class="w"> </span>./<span class="w">                                 </span><span class="sb">`</span><span class="c1"># Output directory` \</span>
-r<span class="w">                                    </span><span class="sb">`</span><span class="c1"># Quantify reporter edits` \</span>
-n<span class="w"> </span><span class="si">${</span><span class="nv">screen_id</span><span class="si">}</span><span class="w">                          </span><span class="sb">`</span><span class="c1"># ID of the screen to be counted`</span>
</pre></div>
</div>
<p>Make sure you follow the <a class="reference external" href="../../README#input-file-format">input file format</a> for seamless downstream steps. This will produce <code class="docutils literal notranslate"><span class="pre">./bean_count_${screen_id}.h5ad</span></code>.</p>
</section>
<section id="qc-samples-guides-qc">
<h3>2. QC samples &amp; guides (<a class="reference internal" href="qc.html#qc"><span class="std std-ref">bean qc</span></a>)<a class="headerlink" href="#qc-samples-guides-qc" title="Link to this heading">¶</a></h3>
<p>Base editing data will include QC about editing efficiency. As QC uses predefined column names and values, beware to follow the <a class="reference external" href="../../README#input-file-format">input file guideline</a>, but you can change the parameters with the full argument list of <cite>``bean-qc`</cite> &lt;../../README#bean-qc-qc-of-reporter-screen-data&gt;`_. (Common factors you may want to tweak is <code class="docutils literal notranslate"><span class="pre">--ctrl-cond=bulk</span></code> and <code class="docutils literal notranslate"><span class="pre">--lfc-conds=top,bot</span></code> if you have different sample condition labels.)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>bean-qc \
  bean_count_${screen_id}.h5ad    `# Input ReporterScreen .h5ad file path` \
  -o bean_count_${screen_id}_masked.h5ad   `# Output ReporterScreen .h5ad file path` \
  -r qc_report_${screen_id}   `# Prefix for QC report`
</pre></div>
</div>
<p>If the data does not include reporter editing data, you can provide <code class="docutils literal notranslate"><span class="pre">--no-editing</span></code> flag to omit the editing rate QC.</p>
</section>
<section id="quantify-variant-effect-run">
<h3>3. Quantify variant effect (<a class="reference internal" href="run.html#run"><span class="std std-ref">bean run</span></a>)<a class="headerlink" href="#quantify-variant-effect-run" title="Link to this heading">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">bean-run</span></code> can take 3 run options to quantify editing rate:</p>
<ol class="arabic">
<li><dl class="simple">
<dt>From <strong>reporter + accessibility</strong><span class="raw-html-m2r"><br></span></dt><dd><p>If your gRNA metadata table (<code class="docutils literal notranslate"><span class="pre">tests/data/test_guide_info.csv</span></code> above) included per-gRNA accessibility score,</p>
</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>   bean-run sorting variant \
   tests/data/bean_count_${screen_id}_masked.h5ad \
   -o tests/test_res/var/ \
   --fit-negctrl \
   --scale-by-acc \
   --accessibility-col accessibility

If your gRNA metadata table (\ ``tests/data/test_guide_info.csv`` above) included per-gRNA chromosome &amp; position and you have bigWig file with accessibility signal,
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>bean-run sorting variant \
tests/data/bean_count_${screen_id}_masked.h5ad \
-o tests/test_res/var/ \
--fit-negctrl \
--scale-by-acc \
--accessibility-bw accessibility.bw
</pre></div>
</div>
</li>
<li><p>From <strong>reporter</strong>, without accessibility</p>
<blockquote>
<div><p>This assumes the all target sites have the uniform chromatin accessibility.</p>
</div></blockquote>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>bean-run sorting variant \
tests/data/bean_count_${screen_id}_masked.h5ad \
-o tests/test_res/var/ \
--fit-negctrl
</pre></div>
</div>
</li>
<li><dl class="simple">
<dt>No reporter information, assume the same editing efficiency of all gRNAs.<span class="raw-html-m2r"><br></span></dt><dd><p>Use this option if your data don’t have editing outcome information.</p>
</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>bean-run sorting variant \
tests/data/bean_count_${screen_id}_masked.h5ad \
-o tests/test_res/var/ \
--fit-negctrl \
--uniform-edit
</pre></div>
</div>
</li>
</ol>
<p>See <a class="reference internal" href="subcommands.html#subcommands"><span class="std std-ref">Subcommands</span></a> for the full details.</p>
</section>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index_.html">bean</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">GWAS variant library</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#variant-sorting-screen-tutorial">Variant sorting screen tutorial</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="cds.html">Coding sequence tiling library</a></li>
<li class="toctree-l1"><a class="reference internal" href="input.html">Input file format</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="subcommands.html">Subcommands</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="ReporterScreen_api.html">ReporterScreen API tutorial</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index_.html">Documentation overview</a><ul>
      <li>Previous: <a href="index_.html" title="previous chapter">Welcome to <cite>bean</cite>’s documentation!</a></li>
      <li>Next: <a href="cds.html" title="next chapter">Coding sequence tiling library</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2024, Jayoung Ryu.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 7.2.6</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 0.7.16</a>
      
      |
      <a href="_sources/gwas.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>