<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>ReporterScreen API tutorial &#8212; bean 1.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=d1102ebc" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=12dfc556" />
    <script src="_static/documentation_options.js?v=8d563738"></script>
    <script src="_static/doctools.js?v=888ff710"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="bean run" href="run.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="reporterscreen-api-tutorial">
<h1>ReporterScreen API tutorial<a class="headerlink" href="#reporterscreen-api-tutorial" title="Link to this heading">¶</a></h1>
<p>Load the required packages. (Anndata import isn’t required to use the package).</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import numpy as np
import pandas as pd
import anndata as ad
import seaborn as sns
import matplotlib.pyplot as plt
import bean as br
</pre></div>
</div>
<p>bean <code class="code highlight bash docutils literal highlight-bash">ReporterScreen</code> object and perturb-seq <code class="code highlight bash docutils literal highlight-bash">Screen</code> object are both <code class="code highlight bash docutils literal highlight-bash">anndata</code> compatible.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = ad.read_h5ad(&quot;bean_count_07+1021_LDLvar.h5ad&quot;)
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 3455 × 12
    obs: &#39;name&#39;, &#39;Unnamed: 0&#39;, &#39;Target gene/variant&#39;, &#39;Target descriptor&#39;, &#39;Arbitrary number&#39;, &#39;gRNA position category&#39;, &#39;Target base position in gRNA&#39;, &#39;Target base position in reporter&#39;, &#39;BE&#39;, &#39;Group&#39;, &#39;sequence&#39;, &#39;Reporter&#39;, &#39;barcode&#39;, &#39;5-nt PAM&#39;, &#39;offset&#39;, &#39;target&#39;, &#39;target_pos&#39;, &#39;Group2&#39;, &#39;masked_sequence&#39;, &#39;masked_barcode&#39;, &#39;edit_rate&#39;
    var: &#39;index&#39;, &#39;sort&#39;, &#39;replicate&#39;
    uns: &#39;allele_counts&#39;, &#39;edit_counts&#39;
    layers: &#39;X_bcmatch&#39;, &#39;edits&#39;
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cdata = br.read_h5ad(&quot;bean_count_07+1021_LDLvar.h5ad&quot;)
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cdata
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Genome</span> <span class="n">Editing</span> <span class="n">Screen</span> <span class="n">comprised</span> <span class="n">of</span> <span class="n">n_guides</span> <span class="n">x</span> <span class="n">n_conditions</span> <span class="o">=</span> <span class="mi">3455</span> <span class="n">x</span> <span class="mi">12</span>
   <span class="n">guides</span><span class="p">:</span>    <span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="s1">&#39;Unnamed: 0&#39;</span><span class="p">,</span> <span class="s1">&#39;Target gene/variant&#39;</span><span class="p">,</span> <span class="s1">&#39;Target descriptor&#39;</span><span class="p">,</span> <span class="s1">&#39;Arbitrary number&#39;</span><span class="p">,</span> <span class="s1">&#39;gRNA position category&#39;</span><span class="p">,</span> <span class="s1">&#39;Target base position in gRNA&#39;</span><span class="p">,</span> <span class="s1">&#39;Target base position in reporter&#39;</span><span class="p">,</span> <span class="s1">&#39;BE&#39;</span><span class="p">,</span> <span class="s1">&#39;Group&#39;</span><span class="p">,</span> <span class="s1">&#39;sequence&#39;</span><span class="p">,</span> <span class="s1">&#39;Reporter&#39;</span><span class="p">,</span> <span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;5-nt PAM&#39;</span><span class="p">,</span> <span class="s1">&#39;offset&#39;</span><span class="p">,</span> <span class="s1">&#39;target&#39;</span><span class="p">,</span> <span class="s1">&#39;target_pos&#39;</span><span class="p">,</span> <span class="s1">&#39;Group2&#39;</span><span class="p">,</span> <span class="s1">&#39;masked_sequence&#39;</span><span class="p">,</span> <span class="s1">&#39;masked_barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;edit_rate&#39;</span>
   <span class="n">samples</span><span class="p">:</span>    <span class="s1">&#39;index&#39;</span><span class="p">,</span> <span class="s1">&#39;sort&#39;</span><span class="p">,</span> <span class="s1">&#39;replicate&#39;</span>
   <span class="n">condit_m</span><span class="p">:</span>
   <span class="n">condit_p</span><span class="p">:</span>
   <span class="n">layers</span><span class="p">:</span>    <span class="s1">&#39;X_bcmatch&#39;</span><span class="p">,</span> <span class="s1">&#39;edits&#39;</span>
   <span class="n">uns</span><span class="p">:</span>       <span class="s1">&#39;allele_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;edit_counts&#39;</span>
</pre></div>
</div>
<ul class="simple">
<li><p><code class="code highlight bash docutils literal highlight-bash">cdata.X</code>: guide count</p></li>
<li><p><code class="code highlight bash docutils literal highlight-bash">cdata.guides</code>: guide metadata</p></li>
<li><p><code class="code highlight bash docutils literal highlight-bash">cdata.samples</code>: sample/condition metadata</p></li>
<li><p><code class="code highlight bash docutils literal highlight-bash">cdata.layers<span class="o">[</span><span class="s2">&quot;X_bcmatch&quot;</span><span class="o">]</span></code>: barcode-matched guide counts</p></li>
<li><p><code class="code highlight bash docutils literal highlight-bash">cdata.layers<span class="o">[</span><span class="s2">&quot;edits&quot;</span><span class="o">]</span></code>: edit counts</p></li>
<li><p><code class="code highlight bash docutils literal highlight-bash">cdata.uns<span class="o">[</span><span class="s2">&quot;allele_counts&quot;</span><span class="o">]</span></code>: allele counts per guide and condition</p></li>
<li><p><code class="code highlight bash docutils literal highlight-bash">cdata.uns<span class="o">[</span><span class="s2">&quot;edit_counts&quot;</span><span class="o">]</span></code>: edit counts per guide and condition</p></li>
</ul>
<p><code class="code highlight bash docutils literal highlight-bash">guides</code> attribute contains the information about each guide.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cdata.guides
</pre></div>
</div>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>Unnamed: 0</th>
      <th>Target gene/variant</th>
      <th>Target descriptor</th>
      <th>Arbitrary number</th>
      <th>gRNA position category</th>
      <th>Target base position in gRNA</th>
      <th>Target base position in reporter</th>
      <th>BE</th>
      <th>Group</th>
      <th>...</th>
      <th>Reporter</th>
      <th>barcode</th>
      <th>5-nt PAM</th>
      <th>offset</th>
      <th>target</th>
      <th>target_pos</th>
      <th>Group2</th>
      <th>masked_sequence</th>
      <th>masked_barcode</th>
      <th>edit_rate</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CONTROL_1_g1</td>
      <td>0</td>
      <td>CONTROL</td>
      <td>NaN</td>
      <td>1</td>
      <td>g1</td>
      <td>4</td>
      <td>10</td>
      <td>ABE</td>
      <td>NegCtrl</td>
      <td>...</td>
      <td>CCAAGCCCTACGCGGTAGGGAACTTTGGGAGC</td>
      <td>GTTT</td>
      <td>GGGAG</td>
      <td>-10</td>
      <td>CONTROL_1</td>
      <td>9</td>
      <td>NegCtrl</td>
      <td>CCTGCGCGGTGGGGGGCTTT</td>
      <td>GTTT</td>
      <td>0.531163</td>
    </tr>
    <tr>
      <th>1</th>
      <td>CONTROL_1_g2</td>
      <td>1</td>
      <td>CONTROL</td>
      <td>NaN</td>
      <td>1</td>
      <td>g2</td>
      <td>5</td>
      <td>11</td>
      <td>ABE</td>
      <td>NegCtrl</td>
      <td>...</td>
      <td>TCCAAGCCCTACGCGGTAGGGAACTTTGGGAG</td>
      <td>AACA</td>
      <td>TGGGA</td>
      <td>-11</td>
      <td>CONTROL_1</td>
      <td>10</td>
      <td>NegCtrl</td>
      <td>CCCTGCGCGGTGGGGGGCTT</td>
      <td>GGCG</td>
      <td>0.640765</td>
    </tr>
    <tr>
      <th>2</th>
      <td>CONTROL_1_g3</td>
      <td>2</td>
      <td>CONTROL</td>
      <td>NaN</td>
      <td>1</td>
      <td>g3</td>
      <td>5</td>
      <td>12</td>
      <td>ABE</td>
      <td>NegCtrl</td>
      <td>...</td>
      <td>GTCCAAGCCCTACGCGGTAGGGAACTTTGGGA</td>
      <td>CGCT</td>
      <td>TTGGG</td>
      <td>-12</td>
      <td>CONTROL_1</td>
      <td>11</td>
      <td>NegCtrl</td>
      <td>CCCTGCGCGGTGGGGGGCT</td>
      <td>CGCT</td>
      <td>0.417709</td>
    </tr>
    <tr>
      <th>3</th>
      <td>CONTROL_1_g4</td>
      <td>3</td>
      <td>CONTROL</td>
      <td>NaN</td>
      <td>1</td>
      <td>g4</td>
      <td>7</td>
      <td>13</td>
      <td>ABE</td>
      <td>NegCtrl</td>
      <td>...</td>
      <td>CGTCCAAGCCCTACGCGGTAGGGAACTTTGGG</td>
      <td>TGAG</td>
      <td>TTTGG</td>
      <td>-13</td>
      <td>CONTROL_1</td>
      <td>12</td>
      <td>NegCtrl</td>
      <td>GGCCCTGCGCGGTGGGGGGC</td>
      <td>TGGG</td>
      <td>0.126400</td>
    </tr>
    <tr>
      <th>4</th>
      <td>CONTROL_1_g5</td>
      <td>4</td>
      <td>CONTROL</td>
      <td>NaN</td>
      <td>1</td>
      <td>g5</td>
      <td>8</td>
      <td>14</td>
      <td>ABE</td>
      <td>NegCtrl</td>
      <td>...</td>
      <td>ACGTCCAAGCCCTACGCGGTAGGGAACTTTGG</td>
      <td>GTAT</td>
      <td>CTTTG</td>
      <td>-14</td>
      <td>CONTROL_1</td>
      <td>13</td>
      <td>NegCtrl</td>
      <td>GGGCCCTGCGCGGTGGGGGG</td>
      <td>GTGT</td>
      <td>0.201104</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>3450</th>
      <td>rs9987289_Maj_ABE_347_g1</td>
      <td>3450</td>
      <td>rs9987289</td>
      <td>Maj</td>
      <td>347</td>
      <td>g1</td>
      <td>3</td>
      <td>10</td>
      <td>ABE</td>
      <td>Variant</td>
      <td>...</td>
      <td>TGCTTGGGCATCAATATCACGTGGAACCAGCC</td>
      <td>CAGT</td>
      <td>CCAGC</td>
      <td>-10</td>
      <td>rs9987289_Maj_ABE_347</td>
      <td>9</td>
      <td>Variant</td>
      <td>GCGTCGGTGTCGCGTGGGG</td>
      <td>CGGT</td>
      <td>0.087379</td>
    </tr>
    <tr>
      <th>3451</th>
      <td>rs9987289_Maj_ABE_347_g2</td>
      <td>3451</td>
      <td>rs9987289</td>
      <td>Maj</td>
      <td>347</td>
      <td>g2</td>
      <td>4</td>
      <td>11</td>
      <td>ABE</td>
      <td>Variant</td>
      <td>...</td>
      <td>ATGCTTGGGCATCAATATCACGTGGAACCAGC</td>
      <td>TCGC</td>
      <td>ACCAG</td>
      <td>-11</td>
      <td>rs9987289_Maj_ABE_347</td>
      <td>10</td>
      <td>Variant</td>
      <td>GGCGTCGGTGTCGCGTGGG</td>
      <td>TCGC</td>
      <td>0.299923</td>
    </tr>
    <tr>
      <th>3452</th>
      <td>rs9987289_Maj_ABE_347_g3</td>
      <td>3452</td>
      <td>rs9987289</td>
      <td>Maj</td>
      <td>347</td>
      <td>g3</td>
      <td>6</td>
      <td>12</td>
      <td>ABE</td>
      <td>Variant</td>
      <td>...</td>
      <td>GATGCTTGGGCATCAATATCACGTGGAACCAG</td>
      <td>GCAC</td>
      <td>AACCA</td>
      <td>-12</td>
      <td>rs9987289_Maj_ABE_347</td>
      <td>11</td>
      <td>Variant</td>
      <td>TGGGCGTCGGTGTCGCGTGG</td>
      <td>GCGC</td>
      <td>0.224973</td>
    </tr>
    <tr>
      <th>3453</th>
      <td>rs9987289_Maj_ABE_347_g4</td>
      <td>3453</td>
      <td>rs9987289</td>
      <td>Maj</td>
      <td>347</td>
      <td>g4</td>
      <td>7</td>
      <td>13</td>
      <td>ABE</td>
      <td>Variant</td>
      <td>...</td>
      <td>AGATGCTTGGGCATCAATATCACGTGGAACCA</td>
      <td>TTGC</td>
      <td>GAACC</td>
      <td>-13</td>
      <td>rs9987289_Maj_ABE_347</td>
      <td>12</td>
      <td>Variant</td>
      <td>TTGGGCGTCGGTGTCGCGTG</td>
      <td>TTGC</td>
      <td>0.265378</td>
    </tr>
    <tr>
      <th>3454</th>
      <td>rs9987289_Maj_ABE_347_g5</td>
      <td>3454</td>
      <td>rs9987289</td>
      <td>Maj</td>
      <td>347</td>
      <td>g5</td>
      <td>8</td>
      <td>14</td>
      <td>ABE</td>
      <td>Variant</td>
      <td>...</td>
      <td>TAGATGCTTGGGCATCAATATCACGTGGAACC</td>
      <td>GCGA</td>
      <td>GGAAC</td>
      <td>-14</td>
      <td>rs9987289_Maj_ABE_347</td>
      <td>13</td>
      <td>Variant</td>
      <td>CTTGGGCGTCGGTGTCGCGT</td>
      <td>GCGG</td>
      <td>0.266573</td>
    </tr>
  </tbody>
</table>
<p>3455 rows × 21 columns</p>
</div><p><code class="code highlight bash docutils literal highlight-bash">samples</code> attribute contains the sample and condition specific information.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cdata.samples
</pre></div>
</div>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>index</th>
      <th>sort</th>
      <th>replicate</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>rep1_bot</td>
      <td>bot</td>
      <td>rep1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>rep2_bot</td>
      <td>bot</td>
      <td>rep2</td>
    </tr>
    <tr>
      <th>2</th>
      <td>rep3_VPA_bot</td>
      <td>bot</td>
      <td>rep3_VPA</td>
    </tr>
    <tr>
      <th>3</th>
      <td>rep4_VPA_bot</td>
      <td>bot</td>
      <td>rep4_VPA</td>
    </tr>
    <tr>
      <th>4</th>
      <td>rep1_bulk</td>
      <td>bulk</td>
      <td>rep1</td>
    </tr>
    <tr>
      <th>5</th>
      <td>rep2_bulk</td>
      <td>bulk</td>
      <td>rep2</td>
    </tr>
    <tr>
      <th>6</th>
      <td>rep3_VPA_bulk</td>
      <td>bulk</td>
      <td>rep3_VPA</td>
    </tr>
    <tr>
      <th>7</th>
      <td>rep4_VPA_bulk</td>
      <td>bulk</td>
      <td>rep4_VPA</td>
    </tr>
    <tr>
      <th>8</th>
      <td>rep1_top</td>
      <td>top</td>
      <td>rep1</td>
    </tr>
    <tr>
      <th>9</th>
      <td>rep2_top</td>
      <td>top</td>
      <td>rep2</td>
    </tr>
    <tr>
      <th>10</th>
      <td>rep3_VPA_top</td>
      <td>top</td>
      <td>rep3_VPA</td>
    </tr>
    <tr>
      <th>11</th>
      <td>rep4_VPA_top</td>
      <td>top</td>
      <td>rep4_VPA</td>
    </tr>
  </tbody>
</table>
</div><p>Allele_counts information is stored in <code class="code highlight bash docutils literal highlight-bash">.uns<span class="o">[</span><span class="s2">&quot;allele_counts&quot;</span><span class="o">]</span></code>.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cdata.uns[&quot;allele_counts&quot;]
</pre></div>
</div>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>guide</th>
      <th>allele</th>
      <th>rep1_bot</th>
      <th>rep2_bot</th>
      <th>rep3_VPA_bot</th>
      <th>rep4_VPA_bot</th>
      <th>rep1_bulk</th>
      <th>rep2_bulk</th>
      <th>rep3_VPA_bulk</th>
      <th>rep4_VPA_bulk</th>
      <th>rep1_top</th>
      <th>rep2_top</th>
      <th>rep3_VPA_top</th>
      <th>rep4_VPA_top</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>12:51779544AGA_Maj_ABE_2_g1</td>
      <td>0:9:+:A&gt;G,5:14:+:A&gt;G</td>
      <td>14</td>
      <td>20</td>
      <td>13</td>
      <td>0</td>
      <td>6</td>
      <td>15</td>
      <td>2</td>
      <td>17</td>
      <td>22</td>
      <td>14</td>
      <td>34</td>
      <td>3</td>
    </tr>
    <tr>
      <th>1</th>
      <td>12:51779544AGA_Maj_ABE_2_g1</td>
      <td>-4:5:+:A&gt;G,-2:7:+:A&gt;G,5:14:+:A&gt;G,10:19:+:A&gt;G</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>12:51779544AGA_Maj_ABE_2_g1</td>
      <td>-7:2:+:A&gt;G,0:9:+:A&gt;G,5:14:+:A&gt;G</td>
      <td>3</td>
      <td>4</td>
      <td>2</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>5</td>
      <td>2</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>12:51779544AGA_Maj_ABE_2_g1</td>
      <td>-9:0:+:G&gt;A,-8:1:+:G&gt;A,-7:2:+:A&gt;C,-6:3:+:C&gt;A,-4...</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>2</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>12:51779544AGA_Maj_ABE_2_g1</td>
      <td>-7:2:+:A&gt;G,10:19:+:A&gt;G</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>438407</th>
      <td>rs9987289_Maj_ABE_347_g5</td>
      <td>4:17:+:A&gt;G,6:19:+:A&gt;G,9:22:+:A&gt;G</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>2</td>
      <td>0</td>
    </tr>
    <tr>
      <th>438408</th>
      <td>rs9987289_Maj_ABE_347_g5</td>
      <td>-12:1:+:A&gt;G,6:19:+:A&gt;G,9:22:+:A&gt;G,11:24:+:G&gt;A</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
    </tr>
    <tr>
      <th>438409</th>
      <td>rs9987289_Maj_ABE_347_g5</td>
      <td>-12:1:+:A&gt;G,6:19:+:A&gt;G,9:22:+:A&gt;G,16:29:+:A&gt;G</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
    </tr>
    <tr>
      <th>438410</th>
      <td>rs9987289_Maj_ABE_347_g5</td>
      <td>-12:1:+:A&gt;G,0:13:+:A&gt;G,6:19:+:A&gt;G,9:22:+:A&gt;G,1...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>438411</th>
      <td>rs9987289_Maj_ABE_347_g5</td>
      <td>-12:1:+:A&gt;G,6:19:+:A&gt;G,9:22:+:A&gt;G,12:25:+:T&gt;G</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
    </tr>
  </tbody>
</table>
<p>438412 rows × 14 columns</p>
</div><p>Base-level edit counts can be saved at <cite>.uns[“edit_counts”]</cite>.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cdata.uns[&quot;edit_counts&quot;]
</pre></div>
</div>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>guide</th>
      <th>edit</th>
      <th>rep1_bot</th>
      <th>rep2_bot</th>
      <th>rep3_VPA_bot</th>
      <th>rep4_VPA_bot</th>
      <th>rep1_bulk</th>
      <th>rep2_bulk</th>
      <th>rep3_VPA_bulk</th>
      <th>rep4_VPA_bulk</th>
      <th>rep1_top</th>
      <th>rep2_top</th>
      <th>rep3_VPA_top</th>
      <th>rep4_VPA_top</th>
      <th>ref_base</th>
      <th>alt_base</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>12:51779544AGA_Maj_ABE_2_g1</td>
      <td>-1:8:+:G&gt;A</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>G</td>
      <td>A</td>
    </tr>
    <tr>
      <th>1</th>
      <td>12:51779544AGA_Maj_ABE_2_g1</td>
      <td>-1:8:+:G&gt;C</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>G</td>
      <td>C</td>
    </tr>
    <tr>
      <th>2</th>
      <td>12:51779544AGA_Maj_ABE_2_g1</td>
      <td>-1:8:+:G&gt;T</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>G</td>
      <td>T</td>
    </tr>
    <tr>
      <th>3</th>
      <td>12:51779544AGA_Maj_ABE_2_g1</td>
      <td>-2:7:+:A&gt;C</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>2</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>A</td>
      <td>C</td>
    </tr>
    <tr>
      <th>4</th>
      <td>12:51779544AGA_Maj_ABE_2_g1</td>
      <td>-2:7:+:A&gt;G</td>
      <td>19</td>
      <td>34</td>
      <td>40</td>
      <td>4</td>
      <td>59</td>
      <td>25</td>
      <td>66</td>
      <td>7</td>
      <td>68</td>
      <td>48</td>
      <td>149</td>
      <td>2</td>
      <td>A</td>
      <td>G</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>217563</th>
      <td>rs9987289_Maj_ABE_347_g5</td>
      <td>8:21:+:C&gt;A</td>
      <td>0</td>
      <td>7</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>C</td>
      <td>A</td>
    </tr>
    <tr>
      <th>217564</th>
      <td>rs9987289_Maj_ABE_347_g5</td>
      <td>8:21:+:C&gt;G</td>
      <td>0</td>
      <td>0</td>
      <td>2</td>
      <td>0</td>
      <td>0</td>
      <td>8</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
      <td>8</td>
      <td>0</td>
      <td>C</td>
      <td>G</td>
    </tr>
    <tr>
      <th>217565</th>
      <td>rs9987289_Maj_ABE_347_g5</td>
      <td>8:21:+:C&gt;T</td>
      <td>0</td>
      <td>0</td>
      <td>7</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>7</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>C</td>
      <td>T</td>
    </tr>
    <tr>
      <th>217566</th>
      <td>rs9987289_Maj_ABE_347_g5</td>
      <td>9:22:+:A&gt;G</td>
      <td>9</td>
      <td>21</td>
      <td>30</td>
      <td>51</td>
      <td>37</td>
      <td>46</td>
      <td>12</td>
      <td>20</td>
      <td>58</td>
      <td>23</td>
      <td>59</td>
      <td>47</td>
      <td>A</td>
      <td>G</td>
    </tr>
    <tr>
      <th>217567</th>
      <td>rs9987289_Maj_ABE_347_g5</td>
      <td>9:22:+:A&gt;T</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>7</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>A</td>
      <td>T</td>
    </tr>
  </tbody>
</table>
<p>217568 rows × 16 columns</p>
</div><section id="subsetting-addition">
<h2>Subsetting &amp; addition<a class="headerlink" href="#subsetting-addition" title="Link to this heading">¶</a></h2>
<p>Works as anndata, supports allele &amp; edit count operations.</p>
<section id="subsetting-selection">
<h3>Subsetting &amp; selection<a class="headerlink" href="#subsetting-selection" title="Link to this heading">¶</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cdata_subset = cdata[:10,cdata.samples.sort == &quot;bulk&quot;]
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="s1">&#39;rep1_bulk&#39;</span><span class="p">,</span> <span class="s1">&#39;rep2_bulk&#39;</span><span class="p">,</span> <span class="s1">&#39;rep3_VPA_bulk&#39;</span><span class="p">,</span> <span class="s1">&#39;rep4_VPA_bulk&#39;</span><span class="p">]</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cdata_subset.uns[&quot;allele_counts&quot;]
</pre></div>
</div>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>guide</th>
      <th>allele</th>
      <th>rep1_bulk</th>
      <th>rep2_bulk</th>
      <th>rep3_VPA_bulk</th>
      <th>rep4_VPA_bulk</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>14979</th>
      <td>CONTROL_10_g1</td>
      <td>-4:5:+:A&gt;G,0:9:+:A&gt;G</td>
      <td>8</td>
      <td>1</td>
      <td>3</td>
      <td>0</td>
    </tr>
    <tr>
      <th>14980</th>
      <td>CONTROL_10_g1</td>
      <td>-7:2:+:C&gt;T</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>10</td>
    </tr>
    <tr>
      <th>14981</th>
      <td>CONTROL_10_g1</td>
      <td>-4:5:+:A&gt;G</td>
      <td>29</td>
      <td>2</td>
      <td>29</td>
      <td>25</td>
    </tr>
    <tr>
      <th>14982</th>
      <td>CONTROL_10_g1</td>
      <td>1:10:+:A&gt;G</td>
      <td>0</td>
      <td>6</td>
      <td>4</td>
      <td>1</td>
    </tr>
    <tr>
      <th>14983</th>
      <td>CONTROL_10_g1</td>
      <td>-4:5:+:A&gt;G,1:10:+:A&gt;G</td>
      <td>1</td>
      <td>11</td>
      <td>5</td>
      <td>12</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>22837</th>
      <td>CONTROL_1_g5</td>
      <td>-13:0:+:A&gt;-,-12:1:+:C&gt;T,-9:4:+:C&gt;G,-8:5:+:C&gt;T,...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>22838</th>
      <td>CONTROL_1_g5</td>
      <td>-6:7:+:A&gt;C,7:20:+:A&gt;G</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>22839</th>
      <td>CONTROL_1_g5</td>
      <td>-13:0:+:A&gt;G,-10:3:+:T&gt;G,0:13:+:A&gt;G,7:20:+:A&gt;G</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>22840</th>
      <td>CONTROL_1_g5</td>
      <td>0:13:+:A&gt;T</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>22841</th>
      <td>CONTROL_1_g5</td>
      <td>0:13:+:A&gt;G,18:31:+:G&gt;A</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
<p>1080 rows × 6 columns</p>
</div></section>
<section id="lfc-calculation-addition">
<h3>LFC calculation &amp; Addition<a class="headerlink" href="#lfc-calculation-addition" title="Link to this heading">¶</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cdata1 = br.read_h5ad(&quot;/data/pinello/PROJECTS/2021_08_ANBE/data/072121_ABE_topbot/bean_counts/LDLvar/032422_crispresso/bean_count_072121_ABE_topbot_LDLvar.h5ad&quot;)
cdata2 = br.read_h5ad(&quot;/data/pinello/PROJECTS/2021_08_ANBE/data/102121_ABE_topbot/bean_counts/LDLvar/032422_crispresso/bean_count_102121_ABE_topbot_LDLvar.h5ad&quot;)
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cdata1.samples[&quot;sort&quot;] = cdata1.samples[&quot;index&quot;].map(lambda s: s.rsplit(&quot;_&quot;, 1)[-1])
cdata1.samples[&quot;replicate&quot;] = cdata1.samples[&quot;index&quot;].map(lambda s: s.rsplit(&quot;_&quot;, 1)[0])
cdata2.samples[&quot;sort&quot;] = cdata2.samples[&quot;index&quot;].map(lambda s: s.rsplit(&quot;_&quot;, 1)[-1])
cdata2.samples[&quot;replicate&quot;] = cdata2.samples[&quot;index&quot;].map(lambda s: s.rsplit(&quot;_&quot;, 1)[0])
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cdata1.log_norm()
lfc1 = cdata1.log_fold_change_reps(&quot;bot&quot;, &quot;top&quot;)
cdata2.log_norm()
lfc2 = cdata2.log_fold_change_reps(&quot;bot&quot;, &quot;top&quot;)
lfcs = lfc1.join(lfc2, lsuffix = &quot;_1&quot;, rsuffix = &quot;_2&quot;)
sns.pairplot(lfcs)
</pre></div>
</div>
<img alt="_images/output_20_2.png" src="_images/output_20_2.png" />
<p>LFC can be aggregated for biological replicates.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cdata1.log_fold_change_aggregate(&quot;bot&quot;, &quot;top&quot;, aggregate_condit = &quot;replicate&quot;)
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cdata1.guides
</pre></div>
</div>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>Unnamed: 0</th>
      <th>Target gene/variant</th>
      <th>Target descriptor</th>
      <th>Arbitrary number</th>
      <th>gRNA position category</th>
      <th>Target base position in gRNA</th>
      <th>Target base position in reporter</th>
      <th>BE</th>
      <th>Group</th>
      <th>...</th>
      <th>Reporter</th>
      <th>barcode</th>
      <th>5-nt PAM</th>
      <th>offset</th>
      <th>target</th>
      <th>target_pos</th>
      <th>Group2</th>
      <th>masked_sequence</th>
      <th>masked_barcode</th>
      <th>bot_top.lfc.median</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CONTROL_1_g1</td>
      <td>0</td>
      <td>CONTROL</td>
      <td>NaN</td>
      <td>1</td>
      <td>g1</td>
      <td>4</td>
      <td>10</td>
      <td>ABE</td>
      <td>NegCtrl</td>
      <td>...</td>
      <td>CCAAGCCCTACGCGGTAGGGAACTTTGGGAGC</td>
      <td>GTTT</td>
      <td>GGGAG</td>
      <td>-10</td>
      <td>CONTROL_1</td>
      <td>9</td>
      <td>NegCtrl</td>
      <td>CCTGCGCGGTGGGGGGCTTT</td>
      <td>GTTT</td>
      <td>-0.158787</td>
    </tr>
    <tr>
      <th>1</th>
      <td>CONTROL_1_g2</td>
      <td>1</td>
      <td>CONTROL</td>
      <td>NaN</td>
      <td>1</td>
      <td>g2</td>
      <td>5</td>
      <td>11</td>
      <td>ABE</td>
      <td>NegCtrl</td>
      <td>...</td>
      <td>TCCAAGCCCTACGCGGTAGGGAACTTTGGGAG</td>
      <td>AACA</td>
      <td>TGGGA</td>
      <td>-11</td>
      <td>CONTROL_1</td>
      <td>10</td>
      <td>NegCtrl</td>
      <td>CCCTGCGCGGTGGGGGGCTT</td>
      <td>GGCG</td>
      <td>-0.212254</td>
    </tr>
    <tr>
      <th>2</th>
      <td>CONTROL_1_g3</td>
      <td>2</td>
      <td>CONTROL</td>
      <td>NaN</td>
      <td>1</td>
      <td>g3</td>
      <td>5</td>
      <td>12</td>
      <td>ABE</td>
      <td>NegCtrl</td>
      <td>...</td>
      <td>GTCCAAGCCCTACGCGGTAGGGAACTTTGGGA</td>
      <td>CGCT</td>
      <td>TTGGG</td>
      <td>-12</td>
      <td>CONTROL_1</td>
      <td>11</td>
      <td>NegCtrl</td>
      <td>CCCTGCGCGGTGGGGGGCT</td>
      <td>CGCT</td>
      <td>0.186679</td>
    </tr>
    <tr>
      <th>3</th>
      <td>CONTROL_1_g4</td>
      <td>3</td>
      <td>CONTROL</td>
      <td>NaN</td>
      <td>1</td>
      <td>g4</td>
      <td>7</td>
      <td>13</td>
      <td>ABE</td>
      <td>NegCtrl</td>
      <td>...</td>
      <td>CGTCCAAGCCCTACGCGGTAGGGAACTTTGGG</td>
      <td>TGAG</td>
      <td>TTTGG</td>
      <td>-13</td>
      <td>CONTROL_1</td>
      <td>12</td>
      <td>NegCtrl</td>
      <td>GGCCCTGCGCGGTGGGGGGC</td>
      <td>TGGG</td>
      <td>-0.022441</td>
    </tr>
    <tr>
      <th>4</th>
      <td>CONTROL_1_g5</td>
      <td>4</td>
      <td>CONTROL</td>
      <td>NaN</td>
      <td>1</td>
      <td>g5</td>
      <td>8</td>
      <td>14</td>
      <td>ABE</td>
      <td>NegCtrl</td>
      <td>...</td>
      <td>ACGTCCAAGCCCTACGCGGTAGGGAACTTTGG</td>
      <td>GTAT</td>
      <td>CTTTG</td>
      <td>-14</td>
      <td>CONTROL_1</td>
      <td>13</td>
      <td>NegCtrl</td>
      <td>GGGCCCTGCGCGGTGGGGGG</td>
      <td>GTGT</td>
      <td>0.457033</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>3450</th>
      <td>rs9987289_Maj_ABE_347_g1</td>
      <td>3450</td>
      <td>rs9987289</td>
      <td>Maj</td>
      <td>347</td>
      <td>g1</td>
      <td>3</td>
      <td>10</td>
      <td>ABE</td>
      <td>Variant</td>
      <td>...</td>
      <td>TGCTTGGGCATCAATATCACGTGGAACCAGCC</td>
      <td>CAGT</td>
      <td>CCAGC</td>
      <td>-10</td>
      <td>rs9987289_Maj_ABE_347</td>
      <td>9</td>
      <td>Variant</td>
      <td>GCGTCGGTGTCGCGTGGGG</td>
      <td>CGGT</td>
      <td>-0.418312</td>
    </tr>
    <tr>
      <th>3451</th>
      <td>rs9987289_Maj_ABE_347_g2</td>
      <td>3451</td>
      <td>rs9987289</td>
      <td>Maj</td>
      <td>347</td>
      <td>g2</td>
      <td>4</td>
      <td>11</td>
      <td>ABE</td>
      <td>Variant</td>
      <td>...</td>
      <td>ATGCTTGGGCATCAATATCACGTGGAACCAGC</td>
      <td>TCGC</td>
      <td>ACCAG</td>
      <td>-11</td>
      <td>rs9987289_Maj_ABE_347</td>
      <td>10</td>
      <td>Variant</td>
      <td>GGCGTCGGTGTCGCGTGGG</td>
      <td>TCGC</td>
      <td>-0.084936</td>
    </tr>
    <tr>
      <th>3452</th>
      <td>rs9987289_Maj_ABE_347_g3</td>
      <td>3452</td>
      <td>rs9987289</td>
      <td>Maj</td>
      <td>347</td>
      <td>g3</td>
      <td>6</td>
      <td>12</td>
      <td>ABE</td>
      <td>Variant</td>
      <td>...</td>
      <td>GATGCTTGGGCATCAATATCACGTGGAACCAG</td>
      <td>GCAC</td>
      <td>AACCA</td>
      <td>-12</td>
      <td>rs9987289_Maj_ABE_347</td>
      <td>11</td>
      <td>Variant</td>
      <td>TGGGCGTCGGTGTCGCGTGG</td>
      <td>GCGC</td>
      <td>-0.339419</td>
    </tr>
    <tr>
      <th>3453</th>
      <td>rs9987289_Maj_ABE_347_g4</td>
      <td>3453</td>
      <td>rs9987289</td>
      <td>Maj</td>
      <td>347</td>
      <td>g4</td>
      <td>7</td>
      <td>13</td>
      <td>ABE</td>
      <td>Variant</td>
      <td>...</td>
      <td>AGATGCTTGGGCATCAATATCACGTGGAACCA</td>
      <td>TTGC</td>
      <td>GAACC</td>
      <td>-13</td>
      <td>rs9987289_Maj_ABE_347</td>
      <td>12</td>
      <td>Variant</td>
      <td>TTGGGCGTCGGTGTCGCGTG</td>
      <td>TTGC</td>
      <td>-0.517138</td>
    </tr>
    <tr>
      <th>3454</th>
      <td>rs9987289_Maj_ABE_347_g5</td>
      <td>3454</td>
      <td>rs9987289</td>
      <td>Maj</td>
      <td>347</td>
      <td>g5</td>
      <td>8</td>
      <td>14</td>
      <td>ABE</td>
      <td>Variant</td>
      <td>...</td>
      <td>TAGATGCTTGGGCATCAATATCACGTGGAACC</td>
      <td>GCGA</td>
      <td>GGAAC</td>
      <td>-14</td>
      <td>rs9987289_Maj_ABE_347</td>
      <td>13</td>
      <td>Variant</td>
      <td>CTTGGGCGTCGGTGTCGCGT</td>
      <td>GCGG</td>
      <td>0.002245</td>
    </tr>
  </tbody>
</table>
<p>3455 rows × 21 columns</p>
</div><p>Technical replicates show decent LFC correlation.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cdata = cdata1 + cdata2
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cdata
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Genome</span> <span class="n">Editing</span> <span class="n">Screen</span> <span class="n">comprised</span> <span class="n">of</span> <span class="n">n_guides</span> <span class="n">x</span> <span class="n">n_conditions</span> <span class="o">=</span> <span class="mi">3455</span> <span class="n">x</span> <span class="mi">12</span>
   <span class="n">guides</span><span class="p">:</span>    <span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="s1">&#39;Unnamed: 0&#39;</span><span class="p">,</span> <span class="s1">&#39;Target gene/variant&#39;</span><span class="p">,</span> <span class="s1">&#39;Target descriptor&#39;</span><span class="p">,</span> <span class="s1">&#39;Arbitrary number&#39;</span><span class="p">,</span> <span class="s1">&#39;gRNA position category&#39;</span><span class="p">,</span> <span class="s1">&#39;Target base position in gRNA&#39;</span><span class="p">,</span> <span class="s1">&#39;Target base position in reporter&#39;</span><span class="p">,</span> <span class="s1">&#39;BE&#39;</span><span class="p">,</span> <span class="s1">&#39;Group&#39;</span><span class="p">,</span> <span class="s1">&#39;sequence&#39;</span><span class="p">,</span> <span class="s1">&#39;Reporter&#39;</span><span class="p">,</span> <span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;5-nt PAM&#39;</span><span class="p">,</span> <span class="s1">&#39;offset&#39;</span><span class="p">,</span> <span class="s1">&#39;target&#39;</span><span class="p">,</span> <span class="s1">&#39;target_pos&#39;</span><span class="p">,</span> <span class="s1">&#39;Group2&#39;</span><span class="p">,</span> <span class="s1">&#39;masked_sequence&#39;</span><span class="p">,</span> <span class="s1">&#39;masked_barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;bot_top.lfc.median&#39;</span>
   <span class="n">samples</span><span class="p">:</span>    <span class="s1">&#39;index&#39;</span><span class="p">,</span> <span class="s1">&#39;sort&#39;</span><span class="p">,</span> <span class="s1">&#39;replicate&#39;</span>
   <span class="n">condit_m</span><span class="p">:</span>
   <span class="n">condit_p</span><span class="p">:</span>
   <span class="n">layers</span><span class="p">:</span>    <span class="s1">&#39;edits&#39;</span><span class="p">,</span> <span class="s1">&#39;X_bcmatch&#39;</span>
   <span class="n">uns</span><span class="p">:</span>       <span class="s1">&#39;allele_counts&#39;</span>
</pre></div>
</div>
<p>You can concatenate different samples with shared guides.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>br.concat((cdata1, cdata2))
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Genome</span> <span class="n">Editing</span> <span class="n">Screen</span> <span class="n">comprised</span> <span class="n">of</span> <span class="n">n_guides</span> <span class="n">x</span> <span class="n">n_conditions</span> <span class="o">=</span> <span class="mi">3455</span> <span class="n">x</span> <span class="mi">24</span>
   <span class="n">guides</span><span class="p">:</span>    <span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="s1">&#39;Unnamed: 0&#39;</span><span class="p">,</span> <span class="s1">&#39;Target gene/variant&#39;</span><span class="p">,</span> <span class="s1">&#39;Target descriptor&#39;</span><span class="p">,</span> <span class="s1">&#39;Arbitrary number&#39;</span><span class="p">,</span> <span class="s1">&#39;gRNA position category&#39;</span><span class="p">,</span> <span class="s1">&#39;Target base position in gRNA&#39;</span><span class="p">,</span> <span class="s1">&#39;Target base position in reporter&#39;</span><span class="p">,</span> <span class="s1">&#39;BE&#39;</span><span class="p">,</span> <span class="s1">&#39;Group&#39;</span><span class="p">,</span> <span class="s1">&#39;sequence&#39;</span><span class="p">,</span> <span class="s1">&#39;Reporter&#39;</span><span class="p">,</span> <span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;5-nt PAM&#39;</span><span class="p">,</span> <span class="s1">&#39;offset&#39;</span><span class="p">,</span> <span class="s1">&#39;target&#39;</span><span class="p">,</span> <span class="s1">&#39;target_pos&#39;</span><span class="p">,</span> <span class="s1">&#39;Group2&#39;</span><span class="p">,</span> <span class="s1">&#39;masked_sequence&#39;</span><span class="p">,</span> <span class="s1">&#39;masked_barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;bot_top.lfc.median&#39;</span>
   <span class="n">samples</span><span class="p">:</span>    <span class="s1">&#39;index&#39;</span><span class="p">,</span> <span class="s1">&#39;sort&#39;</span><span class="p">,</span> <span class="s1">&#39;replicate&#39;</span>
   <span class="n">condit_m</span><span class="p">:</span>
   <span class="n">condit_p</span><span class="p">:</span>
   <span class="n">layers</span><span class="p">:</span>    <span class="s1">&#39;X&#39;</span><span class="p">,</span> <span class="s1">&#39;X_bcmatch&#39;</span><span class="p">,</span> <span class="s1">&#39;edits&#39;</span><span class="p">,</span> <span class="s1">&#39;lognorm_counts&#39;</span><span class="p">,</span> <span class="s1">&#39;lognorm_edits&#39;</span>
   <span class="n">uns</span><span class="p">:</span>       <span class="s1">&#39;allele_counts&#39;</span>
</pre></div>
</div>
</section>
</section>
<section id="getting-edit-rates-from-allele-counts">
<h2>Getting edit rates from allele counts<a class="headerlink" href="#getting-edit-rates-from-allele-counts" title="Link to this heading">¶</a></h2>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cdata.get_edit_rate(normalize_by_editable_base = False,
                   edited_base = &quot;A&quot;,
                   editable_base_start = 3,
                   editable_base_end = 8,
                   bcmatch_thres = 10,
                   prior_weight = 1)
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cdata.uns[&quot;edit_counts&quot;] = cdata.get_edit_from_allele()
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cdata.get_edit_mat_from_uns(&quot;A&quot;, &quot;G&quot;, match_target_position = True)
cdata.get_edit_rate(edited_base = &quot;A&quot;, bcmatch_thres = 10)
plt.hist(cdata.guides.edit_rate, bins=30)
plt.show()
</pre></div>
</div>
<img alt="_images/output_34_1.png" src="_images/output_34_1.png" />
<section id="calculating-lfc">
<h3>Calculating LFC<a class="headerlink" href="#calculating-lfc" title="Link to this heading">¶</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cdata.log_norm()
cdata.log_fold_change_aggregate(&quot;bot&quot;, &quot;top&quot;, aggregate_condit = &quot;replicate&quot;)
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cdata.guides
</pre></div>
</div>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>Unnamed: 0</th>
      <th>Target gene/variant</th>
      <th>Target descriptor</th>
      <th>Arbitrary number</th>
      <th>gRNA position category</th>
      <th>Target base position in gRNA</th>
      <th>Target base position in reporter</th>
      <th>BE</th>
      <th>Group</th>
      <th>...</th>
      <th>barcode</th>
      <th>5-nt PAM</th>
      <th>offset</th>
      <th>target</th>
      <th>target_pos</th>
      <th>Group2</th>
      <th>masked_sequence</th>
      <th>masked_barcode</th>
      <th>bot_top.lfc.median</th>
      <th>edit_rate</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CONTROL_1_g1</td>
      <td>0</td>
      <td>CONTROL</td>
      <td>NaN</td>
      <td>1</td>
      <td>g1</td>
      <td>4</td>
      <td>10</td>
      <td>ABE</td>
      <td>NegCtrl</td>
      <td>...</td>
      <td>GTTT</td>
      <td>GGGAG</td>
      <td>-10</td>
      <td>CONTROL_1</td>
      <td>9</td>
      <td>NegCtrl</td>
      <td>CCTGCGCGGTGGGGGGCTTT</td>
      <td>GTTT</td>
      <td>-0.135550</td>
      <td>0.531163</td>
    </tr>
    <tr>
      <th>1</th>
      <td>CONTROL_1_g2</td>
      <td>1</td>
      <td>CONTROL</td>
      <td>NaN</td>
      <td>1</td>
      <td>g2</td>
      <td>5</td>
      <td>11</td>
      <td>ABE</td>
      <td>NegCtrl</td>
      <td>...</td>
      <td>AACA</td>
      <td>TGGGA</td>
      <td>-11</td>
      <td>CONTROL_1</td>
      <td>10</td>
      <td>NegCtrl</td>
      <td>CCCTGCGCGGTGGGGGGCTT</td>
      <td>GGCG</td>
      <td>-0.059391</td>
      <td>0.640765</td>
    </tr>
    <tr>
      <th>2</th>
      <td>CONTROL_1_g3</td>
      <td>2</td>
      <td>CONTROL</td>
      <td>NaN</td>
      <td>1</td>
      <td>g3</td>
      <td>5</td>
      <td>12</td>
      <td>ABE</td>
      <td>NegCtrl</td>
      <td>...</td>
      <td>CGCT</td>
      <td>TTGGG</td>
      <td>-12</td>
      <td>CONTROL_1</td>
      <td>11</td>
      <td>NegCtrl</td>
      <td>CCCTGCGCGGTGGGGGGCT</td>
      <td>CGCT</td>
      <td>0.141290</td>
      <td>0.417709</td>
    </tr>
    <tr>
      <th>3</th>
      <td>CONTROL_1_g4</td>
      <td>3</td>
      <td>CONTROL</td>
      <td>NaN</td>
      <td>1</td>
      <td>g4</td>
      <td>7</td>
      <td>13</td>
      <td>ABE</td>
      <td>NegCtrl</td>
      <td>...</td>
      <td>TGAG</td>
      <td>TTTGG</td>
      <td>-13</td>
      <td>CONTROL_1</td>
      <td>12</td>
      <td>NegCtrl</td>
      <td>GGCCCTGCGCGGTGGGGGGC</td>
      <td>TGGG</td>
      <td>-0.072358</td>
      <td>0.126400</td>
    </tr>
    <tr>
      <th>4</th>
      <td>CONTROL_1_g5</td>
      <td>4</td>
      <td>CONTROL</td>
      <td>NaN</td>
      <td>1</td>
      <td>g5</td>
      <td>8</td>
      <td>14</td>
      <td>ABE</td>
      <td>NegCtrl</td>
      <td>...</td>
      <td>GTAT</td>
      <td>CTTTG</td>
      <td>-14</td>
      <td>CONTROL_1</td>
      <td>13</td>
      <td>NegCtrl</td>
      <td>GGGCCCTGCGCGGTGGGGGG</td>
      <td>GTGT</td>
      <td>0.269650</td>
      <td>0.201104</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>3450</th>
      <td>rs9987289_Maj_ABE_347_g1</td>
      <td>3450</td>
      <td>rs9987289</td>
      <td>Maj</td>
      <td>347</td>
      <td>g1</td>
      <td>3</td>
      <td>10</td>
      <td>ABE</td>
      <td>Variant</td>
      <td>...</td>
      <td>CAGT</td>
      <td>CCAGC</td>
      <td>-10</td>
      <td>rs9987289_Maj_ABE_347</td>
      <td>9</td>
      <td>Variant</td>
      <td>GCGTCGGTGTCGCGTGGGG</td>
      <td>CGGT</td>
      <td>-0.230264</td>
      <td>0.087379</td>
    </tr>
    <tr>
      <th>3451</th>
      <td>rs9987289_Maj_ABE_347_g2</td>
      <td>3451</td>
      <td>rs9987289</td>
      <td>Maj</td>
      <td>347</td>
      <td>g2</td>
      <td>4</td>
      <td>11</td>
      <td>ABE</td>
      <td>Variant</td>
      <td>...</td>
      <td>TCGC</td>
      <td>ACCAG</td>
      <td>-11</td>
      <td>rs9987289_Maj_ABE_347</td>
      <td>10</td>
      <td>Variant</td>
      <td>GGCGTCGGTGTCGCGTGGG</td>
      <td>TCGC</td>
      <td>-0.182151</td>
      <td>0.299923</td>
    </tr>
    <tr>
      <th>3452</th>
      <td>rs9987289_Maj_ABE_347_g3</td>
      <td>3452</td>
      <td>rs9987289</td>
      <td>Maj</td>
      <td>347</td>
      <td>g3</td>
      <td>6</td>
      <td>12</td>
      <td>ABE</td>
      <td>Variant</td>
      <td>...</td>
      <td>GCAC</td>
      <td>AACCA</td>
      <td>-12</td>
      <td>rs9987289_Maj_ABE_347</td>
      <td>11</td>
      <td>Variant</td>
      <td>TGGGCGTCGGTGTCGCGTGG</td>
      <td>GCGC</td>
      <td>-0.165778</td>
      <td>0.224973</td>
    </tr>
    <tr>
      <th>3453</th>
      <td>rs9987289_Maj_ABE_347_g4</td>
      <td>3453</td>
      <td>rs9987289</td>
      <td>Maj</td>
      <td>347</td>
      <td>g4</td>
      <td>7</td>
      <td>13</td>
      <td>ABE</td>
      <td>Variant</td>
      <td>...</td>
      <td>TTGC</td>
      <td>GAACC</td>
      <td>-13</td>
      <td>rs9987289_Maj_ABE_347</td>
      <td>12</td>
      <td>Variant</td>
      <td>TTGGGCGTCGGTGTCGCGTG</td>
      <td>TTGC</td>
      <td>-0.340590</td>
      <td>0.265378</td>
    </tr>
    <tr>
      <th>3454</th>
      <td>rs9987289_Maj_ABE_347_g5</td>
      <td>3454</td>
      <td>rs9987289</td>
      <td>Maj</td>
      <td>347</td>
      <td>g5</td>
      <td>8</td>
      <td>14</td>
      <td>ABE</td>
      <td>Variant</td>
      <td>...</td>
      <td>GCGA</td>
      <td>GGAAC</td>
      <td>-14</td>
      <td>rs9987289_Maj_ABE_347</td>
      <td>13</td>
      <td>Variant</td>
      <td>CTTGGGCGTCGGTGTCGCGT</td>
      <td>GCGG</td>
      <td>0.034365</td>
      <td>0.266573</td>
    </tr>
  </tbody>
</table>
<p>3455 rows × 22 columns</p>
</div></section>
<section id="allele-translation">
<h3>Allele translation<a class="headerlink" href="#allele-translation" title="Link to this heading">¶</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cdata_tiling = br.read_h5ad(&quot;../../072121_ABE_topbot/bean_counts/LDLRCDS/032422_crispresso/bean_count_072121_ABE_topbot_LDLRCDS.h5ad&quot;)
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cdata_tiling.uns[&quot;allele_counts&quot;].allele
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">0</span>                                         <span class="mi">11224415</span><span class="p">:</span><span class="mi">14</span><span class="p">:</span><span class="o">+</span><span class="p">:</span><span class="n">A</span><span class="o">&gt;</span><span class="n">G</span>
<span class="mi">1</span>                        <span class="mi">11224401</span><span class="p">:</span><span class="mi">0</span><span class="p">:</span><span class="o">+</span><span class="p">:</span><span class="n">A</span><span class="o">&gt;</span><span class="n">G</span><span class="p">,</span><span class="mi">11224415</span><span class="p">:</span><span class="mi">14</span><span class="p">:</span><span class="o">+</span><span class="p">:</span><span class="n">A</span><span class="o">&gt;</span><span class="n">G</span>
<span class="mi">2</span>                        <span class="mi">11224410</span><span class="p">:</span><span class="mi">9</span><span class="p">:</span><span class="o">+</span><span class="p">:</span><span class="n">A</span><span class="o">&gt;</span><span class="n">G</span><span class="p">,</span><span class="mi">11224415</span><span class="p">:</span><span class="mi">14</span><span class="p">:</span><span class="o">+</span><span class="p">:</span><span class="n">A</span><span class="o">&gt;</span><span class="n">G</span>
<span class="mi">3</span>         <span class="mi">11224401</span><span class="p">:</span><span class="mi">0</span><span class="p">:</span><span class="o">+</span><span class="p">:</span><span class="n">A</span><span class="o">&gt;</span><span class="n">G</span><span class="p">,</span><span class="mi">11224402</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="o">+</span><span class="p">:</span><span class="n">A</span><span class="o">&gt;</span><span class="n">G</span><span class="p">,</span><span class="mi">11224410</span><span class="p">:</span><span class="mi">9</span><span class="p">:</span><span class="o">+...</span>
<span class="mi">4</span>                                          <span class="mi">11224401</span><span class="p">:</span><span class="mi">0</span><span class="p">:</span><span class="o">+</span><span class="p">:</span><span class="n">A</span><span class="o">&gt;</span><span class="n">G</span>
                                <span class="o">...</span>
<span class="mi">438001</span>    <span class="mi">11203000</span><span class="p">:</span><span class="mi">4</span><span class="p">:</span><span class="o">+</span><span class="p">:</span><span class="n">A</span><span class="o">&gt;</span><span class="n">G</span><span class="p">,</span><span class="mi">11203002</span><span class="p">:</span><span class="mi">6</span><span class="p">:</span><span class="o">+</span><span class="p">:</span><span class="n">A</span><span class="o">&gt;</span><span class="n">G</span><span class="p">,</span><span class="mi">11203006</span><span class="p">:</span><span class="mi">10</span><span class="p">:</span><span class="o">...</span>
<span class="mi">438002</span>    <span class="mi">11224074</span><span class="p">:</span><span class="mi">0</span><span class="p">:</span><span class="o">+</span><span class="p">:</span><span class="n">A</span><span class="o">&gt;</span><span class="n">G</span><span class="p">,</span><span class="mi">11224086</span><span class="p">:</span><span class="mi">12</span><span class="p">:</span><span class="o">+</span><span class="p">:</span><span class="n">A</span><span class="o">&gt;</span><span class="n">G</span><span class="p">,</span><span class="mi">11224092</span><span class="p">:</span><span class="mf">18.</span><span class="o">..</span>
<span class="mi">438003</span>    <span class="mi">0</span><span class="p">:</span><span class="mi">0</span><span class="p">:</span><span class="o">+</span><span class="p">:</span><span class="n">A</span><span class="o">&gt;</span><span class="n">G</span><span class="p">,</span><span class="mi">3</span><span class="p">:</span><span class="mi">3</span><span class="p">:</span><span class="o">+</span><span class="p">:</span><span class="n">A</span><span class="o">&gt;</span><span class="n">G</span><span class="p">,</span><span class="mi">11</span><span class="p">:</span><span class="mi">11</span><span class="p">:</span><span class="o">+</span><span class="p">:</span><span class="n">A</span><span class="o">&gt;</span><span class="n">G</span><span class="p">,</span><span class="mi">13</span><span class="p">:</span><span class="mi">13</span><span class="p">:</span><span class="o">+</span><span class="p">:</span><span class="n">A</span><span class="o">&gt;</span><span class="n">G</span><span class="p">,</span><span class="mf">17.</span><span class="o">..</span>
<span class="mi">438004</span>                  <span class="mi">11217409</span><span class="p">:</span><span class="mi">23</span><span class="p">:</span><span class="o">+</span><span class="p">:</span><span class="n">G</span><span class="o">&gt;-</span><span class="p">,</span><span class="mi">11217417</span><span class="p">:</span><span class="mi">31</span><span class="p">:</span><span class="o">+</span><span class="p">:</span><span class="o">-&gt;</span><span class="n">C</span>
<span class="mi">438005</span>    <span class="mi">11226735</span><span class="p">:</span><span class="mi">30</span><span class="p">:</span><span class="o">-</span><span class="p">:</span><span class="n">A</span><span class="o">&gt;</span><span class="n">G</span><span class="p">,</span><span class="mi">11226742</span><span class="p">:</span><span class="mi">23</span><span class="p">:</span><span class="o">-</span><span class="p">:</span><span class="n">A</span><span class="o">&gt;</span><span class="n">G</span><span class="p">,</span><span class="mi">11226747</span><span class="p">:</span><span class="mf">1.</span><span class="o">..</span>
<span class="n">Name</span><span class="p">:</span> <span class="n">allele</span><span class="p">,</span> <span class="n">Length</span><span class="p">:</span> <span class="mi">438006</span><span class="p">,</span> <span class="n">dtype</span><span class="p">:</span> <span class="nb">object</span>
</pre></div>
</div>
</section>
<section id="writing">
<h3>Writing<a class="headerlink" href="#writing" title="Link to this heading">¶</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cdata.to_Excel(&quot;tmp.xlsx&quot;)
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Writing</span> <span class="n">to</span><span class="p">:</span> <span class="n">tmp</span><span class="o">.</span><span class="n">xlsx</span>

    <span class="n">Sheet</span> <span class="mi">1</span><span class="p">:</span>        <span class="n">X</span>
    <span class="n">Sheet</span> <span class="mi">2</span><span class="p">:</span>        <span class="n">edits</span>
    <span class="n">Sheet</span> <span class="mi">3</span><span class="p">:</span>        <span class="n">X_bcmatch</span>
    <span class="n">Sheet</span> <span class="mi">4</span><span class="p">:</span>        <span class="n">lognorm_counts</span>
    <span class="n">Sheet</span> <span class="mi">5</span><span class="p">:</span>        <span class="n">lognorm_edits</span>
    <span class="n">Sheet</span> <span class="mi">6</span><span class="p">:</span>        <span class="n">guides</span>
    <span class="n">Sheet</span> <span class="mi">7</span><span class="p">:</span>        <span class="n">samples</span>
    <span class="n">Sheet</span> <span class="mi">8</span><span class="p">:</span>        <span class="n">screen</span><span class="o">.</span><span class="n">uns</span><span class="o">.</span><span class="n">allele_counts</span>
    <span class="n">Sheet</span> <span class="mi">9</span><span class="p">:</span>        <span class="n">screen</span><span class="o">.</span><span class="n">uns</span><span class="o">.</span><span class="n">edit_counts</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cdata.to_mageck_input(&quot;mageck_input.txt&quot;, target_column=&#39;target&#39;)
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>%%bash
head<span class="w"> </span>mageck_input.txt
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sgRNA</span>       <span class="n">gene</span>    <span class="mi">0</span>       <span class="mi">1</span>       <span class="mi">2</span>       <span class="mi">3</span>       <span class="mi">4</span>       <span class="mi">5</span>       <span class="mi">6</span>       <span class="mi">7</span>       <span class="mi">8</span>       <span class="mi">9</span>       <span class="mi">10</span>      <span class="mi">11</span>
<span class="n">CONTROL_1_g1</span>        <span class="n">CONTROL_1</span>       <span class="mi">171</span>     <span class="mi">451</span>     <span class="mi">251</span>     <span class="mi">422</span>     <span class="mi">573</span>     <span class="mi">389</span>     <span class="mi">456</span>     <span class="mi">420</span>     <span class="mi">835</span>     <span class="mi">435</span>     <span class="mi">794</span>     <span class="mi">439</span>
<span class="n">CONTROL_1_g2</span>        <span class="n">CONTROL_1</span>       <span class="mi">145</span>     <span class="mi">278</span>     <span class="mi">257</span>     <span class="mi">206</span>     <span class="mi">364</span>     <span class="mi">273</span>     <span class="mi">389</span>     <span class="mi">254</span>     <span class="mi">527</span>     <span class="mi">498</span>     <span class="mi">768</span>     <span class="mi">195</span>
<span class="n">CONTROL_1_g3</span>        <span class="n">CONTROL_1</span>       <span class="mi">333</span>     <span class="mi">835</span>     <span class="mi">488</span>     <span class="mi">632</span>     <span class="mi">898</span>     <span class="mi">899</span>     <span class="mi">780</span>     <span class="mi">713</span>     <span class="mi">1189</span>    <span class="mi">626</span>     <span class="mi">1146</span>    <span class="mi">603</span>
<span class="n">CONTROL_1_g4</span>        <span class="n">CONTROL_1</span>       <span class="mi">246</span>     <span class="mi">663</span>     <span class="mi">387</span>     <span class="mi">448</span>     <span class="mi">823</span>     <span class="mi">595</span>     <span class="mi">705</span>     <span class="mi">600</span>     <span class="mi">921</span>     <span class="mi">595</span>     <span class="mi">1143</span>    <span class="mi">506</span>
<span class="n">CONTROL_1_g5</span>        <span class="n">CONTROL_1</span>       <span class="mi">243</span>     <span class="mi">647</span>     <span class="mi">434</span>     <span class="mi">529</span>     <span class="mi">776</span>     <span class="mi">451</span>     <span class="mi">700</span>     <span class="mi">676</span>     <span class="mi">1062</span>    <span class="mi">611</span>     <span class="mi">928</span>     <span class="mi">379</span>
<span class="n">CONTROL_10_g1</span>       <span class="n">CONTROL_10</span>      <span class="mi">138</span>     <span class="mi">329</span>     <span class="mi">229</span>     <span class="mi">213</span>     <span class="mi">422</span>     <span class="mi">292</span>     <span class="mi">432</span>     <span class="mi">352</span>     <span class="mi">409</span>     <span class="mi">243</span>     <span class="mi">390</span>     <span class="mi">274</span>
<span class="n">CONTROL_10_g2</span>       <span class="n">CONTROL_10</span>      <span class="mi">187</span>     <span class="mi">468</span>     <span class="mi">402</span>     <span class="mi">479</span>     <span class="mi">643</span>     <span class="mi">369</span>     <span class="mi">428</span>     <span class="mi">469</span>     <span class="mi">796</span>     <span class="mi">422</span>     <span class="mi">787</span>     <span class="mi">404</span>
<span class="n">CONTROL_10_g3</span>       <span class="n">CONTROL_10</span>      <span class="mi">57</span>      <span class="mi">126</span>     <span class="mi">83</span>      <span class="mi">131</span>     <span class="mi">281</span>     <span class="mi">114</span>     <span class="mi">184</span>     <span class="mi">115</span>     <span class="mi">300</span>     <span class="mi">106</span>     <span class="mi">299</span>     <span class="mi">106</span>
<span class="n">CONTROL_10_g4</span>       <span class="n">CONTROL_10</span>      <span class="mi">66</span>      <span class="mi">112</span>     <span class="mi">120</span>     <span class="mi">136</span>     <span class="mi">182</span>     <span class="mi">128</span>     <span class="mi">169</span>     <span class="mi">181</span>     <span class="mi">256</span>     <span class="mi">144</span>     <span class="mi">258</span>     <span class="mi">179</span>
</pre></div>
</div>
</section>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index_.html">bean</a></h1>








<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="gwas.html">GWAS variant library</a></li>
<li class="toctree-l1"><a class="reference internal" href="cds.html">Coding sequence tiling library</a></li>
<li class="toctree-l1"><a class="reference internal" href="input.html">Input file format</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="subcommands.html">Subcommands</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">ReporterScreen API tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#subsetting-addition">Subsetting &amp; addition</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#subsetting-selection">Subsetting &amp; selection</a></li>
<li class="toctree-l3"><a class="reference internal" href="#lfc-calculation-addition">LFC calculation &amp; Addition</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#getting-edit-rates-from-allele-counts">Getting edit rates from allele counts</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#calculating-lfc">Calculating LFC</a></li>
<li class="toctree-l3"><a class="reference internal" href="#allele-translation">Allele translation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#writing">Writing</a></li>
</ul>
</li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index_.html">Documentation overview</a><ul>
      <li>Previous: <a href="run.html" title="previous chapter"><cite>bean run</cite></a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2024, Jayoung Ryu.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 7.2.6</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 0.7.16</a>
      
      |
      <a href="_sources/ReporterScreen_api.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>